<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>3 Spotcalling | masmr: Modular Algorithms for Spotcalling in MERFISH in R</title>
<meta name="author" content="Eugene Kwa">
<meta name="description" content="MERFISH data comprises fluorescent images across different channels and imaging cycles. The desired end result are spot coordinates and spot identities. To that end, we need to perform several...">
<meta name="generator" content="bookdown 0.43.2 with bs4_book()">
<meta property="og:title" content="3 Spotcalling | masmr: Modular Algorithms for Spotcalling in MERFISH in R">
<meta property="og:type" content="book">
<meta property="og:image" content="/./images/masmr.png">
<meta property="og:description" content="MERFISH data comprises fluorescent images across different channels and imaging cycles. The desired end result are spot coordinates and spot identities. To that end, we need to perform several...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="3 Spotcalling | masmr: Modular Algorithms for Spotcalling in MERFISH in R">
<meta name="twitter:description" content="MERFISH data comprises fluorescent images across different channels and imaging cycles. The desired end result are spot coordinates and spot identities. To that end, we need to perform several...">
<meta name="twitter:image" content="/./images/masmr.png">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.9.0/transition.js"></script><script src="libs/bs3compat-0.9.0/tabs.js"></script><script src="libs/bs3compat-0.9.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="libs/bslib-component-js-0.9.0/components.min.js"></script><script src="libs/bslib-component-js-0.9.0/web-components.min.js" type="module"></script><link href="libs/bslib-component-css-0.9.0/components.css" rel="stylesheet">
<script src="libs/bslib-tag-require-0.9.0/tag-require.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">masmr: Modular Algorithms for Spotcalling in MERFISH in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About masmr</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">2</span> Introduction</a></li>
<li><a class="active" href="spotcalling.html"><span class="header-section-number">3</span> Spotcalling</a></li>
<li><a class="" href="image-processing.html"><span class="header-section-number">4</span> Image processing</a></li>
<li><a class="" href="cell-segmentation.html"><span class="header-section-number">5</span> Cell segmentation</a></li>
<li><a class="" href="stitching.html"><span class="header-section-number">6</span> Stitching</a></li>
<li><a class="" href="synthesis.html"><span class="header-section-number">7</span> Synthesis</a></li>
<li><a class="" href="troubleshooting.html"><span class="header-section-number">8</span> Troubleshooting</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="spotcalling" class="section level1" number="3">
<h1>
<span class="header-section-number">3</span> Spotcalling<a class="anchor" aria-label="anchor" href="#spotcalling"><i class="fas fa-link"></i></a>
</h1>
<p>MERFISH data comprises fluorescent images across different channels and imaging cycles. The desired end result are spot coordinates and spot identities.</p>
<p>To that end, we need to perform several subtasks: (1) establish parameters and parse necessary metadata, (2) load and process relevant images, (3) consolidate information across images and decode intensity vectors using the MERFISH codebook, (3) filter out low quality spots, and finally (4) save the output. Subtasks 2-4 are repeatedly performed as we loop through FOVs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">masmr</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div id="pre-processing" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"><i class="fas fa-link"></i></a>
</h2>
<p>Before running your pipeline proper, there are a series of pre-processing steps that each script needs to perform. These will be explained in detail here, but the same principles apply to cell segmentation and consolidation scripts.</p>
<div id="establishing-parameters" class="section level3" number="3.1.1">
<h3>
<span class="header-section-number">3.1.1</span> Establishing parameters<a class="anchor" aria-label="anchor" href="#establishing-parameters"><i class="fas fa-link"></i></a>
</h3>
<p>All scripts should begin by establishing parameters: e.g. location of files to read, location to output files, file formats etc.</p>
<p>This is accomplished in <code>masmr</code> using the <code><a href="https://rdrr.io/pkg/masmr/man/establishParams.html">establishParams()</a></code> function, which automatically creates an environment object called <code>params</code> in the global environment that will be repeatedly referenced by downstream functions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_dir</span> <span class="op">=</span> <span class="st">'C:/Users/kwaje/Downloads/JIN_SIM/20190718_WK_T2_a_B2_R8_12_d50/'</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/establishParams.html">establishParams</a></span><span class="op">(</span></span>
<span>  </span>
<span>  <span class="co">## User to fill in</span></span>
<span>  imageDirs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">'S10/'</span><span class="op">)</span>, <span class="co"># Where your images are</span></span>
<span>  imageFormat <span class="op">=</span> <span class="st">'.ome.tif'</span>, <span class="co"># What format your images are in</span></span>
<span>  outDir <span class="op">=</span> <span class="st">'C:/Users/kwaje/Downloads/JYOUT/'</span>, <span class="co"># Where to send output to</span></span>
<span>  codebookFileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>,<span class="st">'codebook/'</span><span class="op">)</span>,pattern<span class="op">=</span><span class="st">'codebook'</span>,full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="co"># Filename and location of your codebook</span></span>
<span>  dapiDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">'DAPI_1/'</span><span class="op">)</span>, <span class="co"># Where DAPI images are</span></span>
<span>  </span>
<span>  <span class="co">## Will attempt automatic fill</span></span>
<span>  metaFormat <span class="op">=</span> <span class="st">'.ome.tif'</span>, <span class="co"># Depends on imageFormat: assumd </span></span>
<span>  dapiFormat <span class="op">=</span> <span class="st">'.ome.tif'</span>, <span class="co"># Assumed to be the same as imageFormat above (but written it out here explicitly)</span></span>
<span>  codebookColumnNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span>,<span class="st">'_'</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'cy3'</span>,<span class="st">'alexa594'</span>,<span class="st">'cy5'</span>,<span class="st">'cy7'</span><span class="op">)</span>,each<span class="op">=</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span>, <span class="co">#If not provided, will attempt placeholder names</span></span>
<span>  pythonLocation <span class="op">=</span> <span class="st">'C:/Users/kwaje/AppData/Local/miniforge3/envs/scipyenv/'</span>, <span class="co">#If unspecified, defaults to reticulate::miniconda_path()</span></span>
<span>  </span>
<span>  <span class="co">## Automatically filled</span></span>
<span>  seed <span class="op">=</span> <span class="fl">12345</span>, <span class="co"># The default seed is 12345</span></span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># It is generally a good idea to keep verbose as True, but will turn it off for this showcase</span></span>
<span>  resumeMode <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># Set this to FALSE if troubleshooting, to overwrite existing files. TRUE for resuming from last checkpoint. </span></span>
<span>  </span>
<span>  <span class="co">## Optional (but recommended for QC checks)</span></span>
<span>  fpkmFileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"FPKM_data/Jin_FPKMData_B.tsv"</span><span class="op">)</span>,</span>
<span>  nProbesFileName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"ProbesPerGene.csv"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As the bare minimum, users need to establish:</p>
<ul>
<li>
<code>imageDirs</code>: where the MERFISH image files are located.</li>
<li>
<code>imageFormat</code>: what image file formats to read. Typically <code>.ome.tif</code> or <code>.dax</code>, but other formats compatible with Bioformats are acceptable too.</li>
<li>
<code>codebookFileName</code>: location of MERFISH codebook.</li>
<li>
<code>codebookColumnNames</code>: names of every bit in the MERFISH codebook. Each “bit name” should typically comprise the image index (e.g. image 1, 2, 3 etc) and channel (e.g. Cy3, Alexa-594 etc): so a bit titled <code>0_cy3</code> tells us that the image to be loaded is the Cy3 channel of the first image in the set (here, starting from 0). This is important because the order images are loaded in may not match up with the order of bits in the codebook.</li>
<li>
<code>outDir</code>: where output files should be written to.</li>
</ul>
<p>Other parameters such as <code>dapiDir</code> and <code>pythonLocation</code> will be discussed subsequently for Cell Segmentation.</p>
<p>There also exist some parameters such as <code>fpkmFileName</code> (file containing FPKM data) and <code>nProbesFileName</code> (file containing number of MERFISH encoding probes per gene), which are files to be read for quality control purposes. However, these parameters are designed to be optional, and your code should run fine even if they are not specified.</p>
<p>The information stored in the <code>params</code> object can be read for further details:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span> envir <span class="op">=</span> <span class="va">params</span> <span class="op">)</span></span></code></pre></div>
<pre><code>##  [1] "codebook_colnames"         "codebook_file"             "dapi_dir"                  "dapi_format"               "dir_names"                
##  [6] "fov_names"                 "fpkm_file"                 "im_dir"                    "im_format"                 "inferred_codebookColnames"
## [11] "meta_format"               "n_probes_file"             "nbits"                     "out_dir"                   "parent_out_dir"           
## [16] "python_location"           "raw_codebook"              "resumeMode"                "seed"                      "verbose"</code></pre>
<p>Note the existence of the <code>params$resumeMode</code> object. This houses a boolean that determines if functions are skipped if there is evidence that they have been run before. As default, to save time, this is set to <code>TRUE</code>.</p>
<div class="card bslib-card bslib-mb-spacing html-fill-item html-fill-container" data-bslib-card-init data-require-bs-caller="card()" data-require-bs-version="5">
<div class="card-header">
<p><strong>WARNING!</strong></p>
</div>
<div class="card-body bslib-gap-spacing html-fill-item html-fill-container" style="margin-top:auto;margin-bottom:auto;flex:1 1 auto;">
<p>When troubleshooting, it is advised that resume mode be turned <strong>off</strong> so that any intermediate files can be overwritten (otherwise, pipeline will resume from those intermediate files). Alternatively, users can leave resume mode <strong>on</strong> but go into the output directory and <strong>manually delete any files they wish to overwrite</strong>.</p>
</div>
<script data-bslib-card-init>bslib.Card.initializeAllCards();</script>
</div>
</div>
<div id="getting-parameters-from-image-metadata" class="section level3" number="3.1.2">
<h3>
<span class="header-section-number">3.1.2</span> Getting parameters from image metadata<a class="anchor" aria-label="anchor" href="#getting-parameters-from-image-metadata"><i class="fas fa-link"></i></a>
</h3>
<p>The next step is to obtain image metadata that accompanies most microscopy images. This is typically found in the same file as your image data (e.g. <code>.ome.tif</code>) but can sometimes be in a separate file (e.g. <code>.xml</code> that accompany <code>.dax</code> images). <code>masmr</code> will attempt to guess which meta data format is appropriate, but users can change values in <code>params</code> as needed.</p>
<p>We can check what values have been entered into <code>params</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">params</span><span class="op">$</span><span class="va">meta_format</span></span></code></pre></div>
<pre><code>## [1] ".ome.tif"</code></pre>
<p>If unsatisfied, we can update as follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">params</span><span class="op">$</span><span class="va">meta_format</span> <span class="op">&lt;-</span> <span class="st">'.dax'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">meta_format</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] ".dax"</code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">params</span><span class="op">$</span><span class="va">meta_format</span> <span class="op">&lt;-</span> <span class="st">'.ome.tif'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">meta_format</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] ".ome.tif"</code></pre>
<p>Having specified our meta data formats, we can then load additional meta data information with the <code>readImageMetaData</code> function.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/readImageMetaData.html">readImageMetaData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Warning in readImageMetaData(): Assuming this data has been acquired by Triton</code></pre>
<pre><code>## Warning in readImageMetaData(): Updating params$fov_names</code></pre>
<p>This adds additional information to <code>params</code>, such as resolution information:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">resolutions</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## $per_pixel_microns
## [1] "0.0706" "0.0706"
## 
## $xydimensions_pixels
## [1] "2960" "2960"
## 
## $xydimensions_microns
## [1] "208.976" "208.976"
## 
## $channel_order
## [1] "cy7"      "cy5"      "alexa594" "cy3"     
## 
## $fov_names
##   [1] "MMStack_1-Pos_000_000" "MMStack_1-Pos_001_000" "MMStack_1-Pos_002_000" "MMStack_1-Pos_003_000" "MMStack_1-Pos_004_000" "MMStack_1-Pos_005_000"
##   [7] "MMStack_1-Pos_006_000" "MMStack_1-Pos_007_000" "MMStack_1-Pos_008_000" "MMStack_1-Pos_009_000" "MMStack_1-Pos_009_001" "MMStack_1-Pos_008_001"
##  [13] "MMStack_1-Pos_007_001" "MMStack_1-Pos_006_001" "MMStack_1-Pos_005_001" "MMStack_1-Pos_004_001" "MMStack_1-Pos_003_001" "MMStack_1-Pos_002_001"
##  [19] "MMStack_1-Pos_001_001" "MMStack_1-Pos_000_001" "MMStack_1-Pos_000_002" "MMStack_1-Pos_001_002" "MMStack_1-Pos_002_002" "MMStack_1-Pos_003_002"
##  [25] "MMStack_1-Pos_004_002" "MMStack_1-Pos_005_002" "MMStack_1-Pos_006_002" "MMStack_1-Pos_007_002" "MMStack_1-Pos_008_002" "MMStack_1-Pos_009_002"
##  [31] "MMStack_1-Pos_009_003" "MMStack_1-Pos_008_003" "MMStack_1-Pos_007_003" "MMStack_1-Pos_006_003" "MMStack_1-Pos_005_003" "MMStack_1-Pos_004_003"
##  [37] "MMStack_1-Pos_003_003" "MMStack_1-Pos_002_003" "MMStack_1-Pos_001_003" "MMStack_1-Pos_000_003" "MMStack_1-Pos_000_004" "MMStack_1-Pos_001_004"
##  [43] "MMStack_1-Pos_002_004" "MMStack_1-Pos_003_004" "MMStack_1-Pos_004_004" "MMStack_1-Pos_005_004" "MMStack_1-Pos_006_004" "MMStack_1-Pos_007_004"
##  [49] "MMStack_1-Pos_008_004" "MMStack_1-Pos_009_004" "MMStack_1-Pos_009_005" "MMStack_1-Pos_008_005" "MMStack_1-Pos_007_005" "MMStack_1-Pos_006_005"
##  [55] "MMStack_1-Pos_005_005" "MMStack_1-Pos_004_005" "MMStack_1-Pos_003_005" "MMStack_1-Pos_002_005" "MMStack_1-Pos_001_005" "MMStack_1-Pos_000_005"
##  [61] "MMStack_1-Pos_000_006" "MMStack_1-Pos_001_006" "MMStack_1-Pos_002_006" "MMStack_1-Pos_003_006" "MMStack_1-Pos_004_006" "MMStack_1-Pos_005_006"
##  [67] "MMStack_1-Pos_006_006" "MMStack_1-Pos_007_006" "MMStack_1-Pos_008_006" "MMStack_1-Pos_009_006" "MMStack_1-Pos_009_007" "MMStack_1-Pos_008_007"
##  [73] "MMStack_1-Pos_007_007" "MMStack_1-Pos_006_007" "MMStack_1-Pos_005_007" "MMStack_1-Pos_004_007" "MMStack_1-Pos_003_007" "MMStack_1-Pos_002_007"
##  [79] "MMStack_1-Pos_001_007" "MMStack_1-Pos_000_007" "MMStack_1-Pos_000_008" "MMStack_1-Pos_001_008" "MMStack_1-Pos_002_008" "MMStack_1-Pos_003_008"
##  [85] "MMStack_1-Pos_004_008" "MMStack_1-Pos_005_008" "MMStack_1-Pos_006_008" "MMStack_1-Pos_007_008" "MMStack_1-Pos_008_008" "MMStack_1-Pos_009_008"
##  [91] "MMStack_1-Pos_009_009" "MMStack_1-Pos_008_009" "MMStack_1-Pos_007_009" "MMStack_1-Pos_006_009" "MMStack_1-Pos_005_009" "MMStack_1-Pos_004_009"
##  [97] "MMStack_1-Pos_003_009" "MMStack_1-Pos_002_009" "MMStack_1-Pos_001_009" "MMStack_1-Pos_000_009"
## 
## $zslices
## [1] "1"</code></pre>
<p>It also creates a dataframe that stores file information for every bit. This will be referenced heavily by downstream functions.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">global_coords</span> <span class="op">)</span>, format <span class="op">=</span> <span class="st">'html'</span>, booktabs <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-table"><table class="table table-sm">
<thead><tr>
<th style="text-align:right;">
x_microns
</th>
<th style="text-align:right;">
y_microns
</th>
<th style="text-align:right;">
z_microns
</th>
<th style="text-align:left;">
image_file
</th>
<th style="text-align:left;">
fov
</th>
<th style="text-align:left;">
bit_name_full
</th>
<th style="text-align:left;">
bit_name_alias
</th>
<th style="text-align:left;">
bit_name
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
-2645.753
</td>
<td style="text-align:right;">
3421.847
</td>
<td style="text-align:right;">
5673.78
</td>
<td style="text-align:left;">
C:/Users/kwaje/Downloads/JIN_SIM/20190718_WK_T2_a_B2_R8_12_d50/S10/hyb00_1/hyb00_1_MMStack_1-Pos_000_000.ome.tif
</td>
<td style="text-align:left;">
MMStack_1-Pos_000_000
</td>
<td style="text-align:left;">
hyb00_1_cy7
</td>
<td style="text-align:left;">
0_cy7
</td>
<td style="text-align:left;">
0_cy7
</td>
</tr>
<tr>
<td style="text-align:right;">
-2645.753
</td>
<td style="text-align:right;">
3421.847
</td>
<td style="text-align:right;">
5673.78
</td>
<td style="text-align:left;">
C:/Users/kwaje/Downloads/JIN_SIM/20190718_WK_T2_a_B2_R8_12_d50/S10/hyb00_1/hyb00_1_MMStack_1-Pos_000_000.ome.tif
</td>
<td style="text-align:left;">
MMStack_1-Pos_000_000
</td>
<td style="text-align:left;">
hyb00_1_cy5
</td>
<td style="text-align:left;">
0_cy5
</td>
<td style="text-align:left;">
0_cy5
</td>
</tr>
<tr>
<td style="text-align:right;">
-2645.753
</td>
<td style="text-align:right;">
3421.847
</td>
<td style="text-align:right;">
5673.78
</td>
<td style="text-align:left;">
C:/Users/kwaje/Downloads/JIN_SIM/20190718_WK_T2_a_B2_R8_12_d50/S10/hyb00_1/hyb00_1_MMStack_1-Pos_000_000.ome.tif
</td>
<td style="text-align:left;">
MMStack_1-Pos_000_000
</td>
<td style="text-align:left;">
hyb00_1_alexa594
</td>
<td style="text-align:left;">
0_alexa594
</td>
<td style="text-align:left;">
0_alexa594
</td>
</tr>
<tr>
<td style="text-align:right;">
-2645.753
</td>
<td style="text-align:right;">
3421.847
</td>
<td style="text-align:right;">
5673.78
</td>
<td style="text-align:left;">
C:/Users/kwaje/Downloads/JIN_SIM/20190718_WK_T2_a_B2_R8_12_d50/S10/hyb00_1/hyb00_1_MMStack_1-Pos_000_000.ome.tif
</td>
<td style="text-align:left;">
MMStack_1-Pos_000_000
</td>
<td style="text-align:left;">
hyb00_1_cy3
</td>
<td style="text-align:left;">
0_cy3
</td>
<td style="text-align:left;">
0_cy3
</td>
</tr>
<tr>
<td style="text-align:right;">
-2645.753
</td>
<td style="text-align:right;">
3421.847
</td>
<td style="text-align:right;">
5673.78
</td>
<td style="text-align:left;">
C:/Users/kwaje/Downloads/JIN_SIM/20190718_WK_T2_a_B2_R8_12_d50/S10/hyb01_1/hyb01_1_MMStack_1-Pos_000_000.ome.tif
</td>
<td style="text-align:left;">
MMStack_1-Pos_000_000
</td>
<td style="text-align:left;">
hyb01_1_cy7
</td>
<td style="text-align:left;">
1_cy7
</td>
<td style="text-align:left;">
1_cy7
</td>
</tr>
<tr>
<td style="text-align:right;">
-2645.753
</td>
<td style="text-align:right;">
3421.847
</td>
<td style="text-align:right;">
5673.78
</td>
<td style="text-align:left;">
C:/Users/kwaje/Downloads/JIN_SIM/20190718_WK_T2_a_B2_R8_12_d50/S10/hyb01_1/hyb01_1_MMStack_1-Pos_000_000.ome.tif
</td>
<td style="text-align:left;">
MMStack_1-Pos_000_000
</td>
<td style="text-align:left;">
hyb01_1_cy5
</td>
<td style="text-align:left;">
1_cy5
</td>
<td style="text-align:left;">
1_cy5
</td>
</tr>
</tbody>
</table></div>
<p>Note that as default, the above information is saved to a sub-directory in your chosen output folder.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">parent_out_dir</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "C:/Users/kwaje/Downloads/JYOUT/"</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">out_dir</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "C:/Users/kwaje/Downloads/JYOUT//IM/"</code></pre>
</div>
<div id="preparing-the-codebook" class="section level3" number="3.1.3">
<h3>
<span class="header-section-number">3.1.3</span> Preparing the codebook<a class="anchor" aria-label="anchor" href="#preparing-the-codebook"><i class="fas fa-link"></i></a>
</h3>
<p>A codebook is a table of indicating the sequence of ON (1) and OFF (0) bits for each gene (example below). {masmr} expects this to be a table (e.g. <code>.csv</code> or <code>.txt</code>) that contains N rows for N genes/blanks, and M+1 columns for M bits (e.g. below is a 16-bit setup) and 1 gene column.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">codebook_file</span>, data.table <span class="op">=</span> <span class="cn">F</span> <span class="op">)</span><span class="op">)</span>, format <span class="op">=</span> <span class="st">'html'</span>, booktabs <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-table"><table class="table table-sm">
<thead><tr>
<th style="text-align:right;">
V1
</th>
<th style="text-align:right;">
V2
</th>
<th style="text-align:right;">
V3
</th>
<th style="text-align:right;">
V4
</th>
<th style="text-align:right;">
V5
</th>
<th style="text-align:right;">
V6
</th>
<th style="text-align:right;">
V7
</th>
<th style="text-align:right;">
V8
</th>
<th style="text-align:right;">
V9
</th>
<th style="text-align:right;">
V10
</th>
<th style="text-align:right;">
V11
</th>
<th style="text-align:right;">
V12
</th>
<th style="text-align:right;">
V13
</th>
<th style="text-align:right;">
V14
</th>
<th style="text-align:right;">
V15
</th>
<th style="text-align:right;">
V16
</th>
<th style="text-align:left;">
V17
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Blank-1
</td>
</tr>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Blank-2
</td>
</tr>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Blank-3
</td>
</tr>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Blank-4
</td>
</tr>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Blank-5
</td>
</tr>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
ASCL1
</td>
</tr>
</tbody>
</table></div>
<p>Unique to spotcalling, users must “prepare” their codebooks. This typically involves ensuring that the order of bits matches the order of image loading that <code>masmr</code> performs. Additionally, <code>masmr</code> also provides the option to add more “blanks” beyond those already specified by the user.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/prepareCodebook.html">prepareCodebook</a></span><span class="op">(</span> exhaustiveBlanks <span class="op">=</span> <span class="cn">T</span> <span class="op">)</span></span></code></pre></div>
<p>The new ordered codebook is stored in the output directory, and users can refer to it as follows:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">ordered_codebook</span> <span class="op">)</span>, format <span class="op">=</span> <span class="st">'html'</span>, booktabs <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-table"><table class="table table-sm">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
0_cy7
</th>
<th style="text-align:right;">
0_cy5
</th>
<th style="text-align:right;">
0_alexa594
</th>
<th style="text-align:right;">
0_cy3
</th>
<th style="text-align:right;">
1_cy7
</th>
<th style="text-align:right;">
1_cy5
</th>
<th style="text-align:right;">
1_alexa594
</th>
<th style="text-align:right;">
1_cy3
</th>
<th style="text-align:right;">
2_cy7
</th>
<th style="text-align:right;">
2_cy5
</th>
<th style="text-align:right;">
2_alexa594
</th>
<th style="text-align:right;">
2_cy3
</th>
<th style="text-align:right;">
3_cy7
</th>
<th style="text-align:right;">
3_cy5
</th>
<th style="text-align:right;">
3_alexa594
</th>
<th style="text-align:right;">
3_cy3
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Blank-1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Blank-2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Blank-3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Blank-4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Blank-5
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
ASCL1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table></div>
</div>
<div id="house-keeping" class="section level3" number="3.1.4">
<h3>
<span class="header-section-number">3.1.4</span> House-keeping<a class="anchor" aria-label="anchor" href="#house-keeping"><i class="fas fa-link"></i></a>
</h3>
<p>Having performed all the necessary pre-processing, it may be useful for users to take a snapshot of the current environment. Unnecessary data can then be cleared from memory in subsequent steps to return the global environment to that snapshot. This is accomplished using our <code>establishCleanSlate</code> and <code>cleanUp</code> functions.</p>
<p>An example use-case is provided below:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/establishCleanSlate.html">establishCleanSlate</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] ".Random.seed" "data_dir"     "full_res"     "n_channels"   "params"</code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">cleanSlate</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] ".Random.seed" "data_dir"     "full_res"     "n_channels"   "params"       "cleanSlate"</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">=</span> <span class="st">'12345'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/exists.html">exists</a></span><span class="op">(</span><span class="st">'x'</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/cleanUp.html">cleanUp</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/exists.html">exists</a></span><span class="op">(</span><span class="st">'x'</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
</div>
</div>
<div id="preparing-to-loop-through-images" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> Preparing to loop through images<a class="anchor" aria-label="anchor" href="#preparing-to-loop-through-images"><i class="fas fa-link"></i></a>
</h2>
<p>Having done the necessary pre-processing, users may now begin processing each FOV. This is done by looping through <code>params$fov_names</code>. It is also recommended that the current FOV be saved in <code>params$current_fov</code>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">fov_names</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "MMStack_1-Pos_000_000" "MMStack_1-Pos_001_000" "MMStack_1-Pos_002_000" "MMStack_1-Pos_003_000" "MMStack_1-Pos_004_000" "MMStack_1-Pos_005_000"</code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">params</span><span class="op">$</span><span class="va">current_fov</span> <span class="op">&lt;-</span> <span class="va">params</span><span class="op">$</span><span class="va">fov_names</span><span class="op">[</span><span class="fl">38</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">current_fov</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "MMStack_1-Pos_002_003"</code></pre>
<p>Users may also wish to check if they are ready to loop with our <code>checkReadyLoop</code> function, but this is entirely optional.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/checkReadyLoop.html">checkReadyLoop</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Warning in checkReadyLoop(): params$brightness_min not found:
##       If desired, ensure getAnchorParams() has been run</code></pre>
<pre><code>## [1] FALSE</code></pre>
<div id="image-reading" class="section level3" number="3.2.1">
<h3>
<span class="header-section-number">3.2.1</span> Image reading<a class="anchor" aria-label="anchor" href="#image-reading"><i class="fas fa-link"></i></a>
</h3>
<p>To read images, we rely on the <code>RBioformats</code> package, which is able to read most image file formats. One exception is <code>.dax</code> files – however, these are binary files and can be read by base R. To account for these different types of images, we can use the <code>readImage</code> function in <code>masmr</code>.</p>
<p>Below, I provide an example of a <code>.dax</code> file:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file</span> <span class="op">&lt;-</span> <span class="st">'C:/Users/kwaje/Downloads/LIU_CONNECT_CORTICALORGANOID_240902/DATA/20240902/cont/Alexa594_0_10.dax'</span></span>
<span><span class="va">im</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/readImage.html">readImage</a></span><span class="op">(</span> </span>
<span>  <span class="va">file</span>,</span>
<span>  imageDimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2048</span>,<span class="fl">2048</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  nchannels <span class="op">=</span> <span class="fl">1</span>, <span class="co">#The image only has 1 channel</span></span>
<span>  nzs <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code>## Warning in readImage(file, imageDimensions = c(1, 2048, 2048, 1), nchannels = 1, : Unable to read with RBioFormats</code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">im</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imsub</span> <span class="op">&lt;-</span> <span class="va">im</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span><span class="va">imsub</span><span class="op">)</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'Image dimensions: '</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">imsub</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">' x '</span><span class="op">)</span> <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-imageread-dax-1.png" width="672"></div>
<p>From above, note that the default <code>imageDimensions</code> is assumed to be: Number of Channels, Number of rows, Number of Columns, Number of Z slices. In the above case, since <code>nchannels</code> and <code>nzs</code> are both 1, users could technically specify image dimensions using width an height alone. However, it is generally good practice to keep to a consistent format.</p>
<p>For <code>.ome.tif</code> files, the same function can be used. However, we need to specify an additional <code>series</code> parameter: otherwise, all FOVs will be loaded. For more details, we recommend reading the documentation for <code>RBioformats::read.image</code>. Below, I provide an example of reading the second channel for the current FOV (specified above).</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">global_coords</span><span class="op">[</span><span class="va">params</span><span class="op">$</span><span class="va">global_coords</span><span class="op">$</span><span class="va">fov</span><span class="op">==</span><span class="va">params</span><span class="op">$</span><span class="va">fov_names</span>, <span class="st">'image_file'</span><span class="op">]</span> <span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">series</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">fov_names</span> <span class="op">==</span> <span class="va">params</span><span class="op">$</span><span class="va">current_fov</span> <span class="op">)</span></span>
<span><span class="va">im</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/readImage.html">readImage</a></span><span class="op">(</span> </span>
<span>  <span class="va">file</span>, </span>
<span>  nchannels <span class="op">=</span> <span class="fl">4</span>, <span class="co"># The image has 4 channels total</span></span>
<span>  nzs <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  series <span class="op">=</span> <span class="va">series</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">im</span><span class="op">)</span> <span class="op">)</span> <span class="co">#4 channels</span></span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imsub</span> <span class="op">&lt;-</span> <span class="va">im</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span> <span class="va">imsub</span> <span class="op">)</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'Image dimensions: '</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">imsub</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">' x '</span><span class="op">)</span> <span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## Warning in as.cimg.array(imsub): Assuming third dimension corresponds to time/depth</code></pre>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-imageread-ometif-1.png" width="672"></div>
<p>Note that <code>readImage</code> as written above reads all four channels into memory, and we subset the second channel using <code>im[,,2]</code>. An alternative is to just load the second channel directly using the <code>subset</code> parameter in <code>RBioformats::read.image</code>, or the <code>channelIndex</code> parameter in <code>readImage</code>. This is useful for memory management.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">im</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/readImage.html">readImage</a></span><span class="op">(</span> </span>
<span>  <span class="va">file</span>,</span>
<span>  nchannels <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  nzs <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  series <span class="op">=</span> <span class="va">series</span>, </span>
<span>  channelIndex <span class="op">=</span> <span class="fl">2</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">im</span><span class="op">)</span><span class="op">)</span> <span class="co">#Only 1 channel loaded</span></span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imsub</span> <span class="op">&lt;-</span> <span class="va">im</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span><span class="va">imsub</span><span class="op">)</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'Image dimensions: '</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">imsub</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">' x '</span><span class="op">)</span> <span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## Warning in as.cimg.array(imsub): Assuming third dimension corresponds to time/depth</code></pre>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-imageread-onechannel-1.png" width="672"></div>
<p>For MERFISH, we typically have to handle multiple images at once. We have thus created an additional image reading function called <code>readImageList</code> that loads all images for a given FOV, by cross-referencing <code>params$global_coords</code> to find all relevant files. And because <code>readImageList</code> does this cross-referencing, parameters such as <code>nchannels</code>, <code>nzs</code> etc do not have to be specified.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/readImageList.html">readImageList</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">current_fov</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">imList</span><span class="op">)</span></span></code></pre></div>
<pre><code>##  [1] "0_cy7"      "0_cy5"      "0_alexa594" "0_cy3"      "1_cy7"      "1_cy5"      "1_alexa594" "1_cy3"      "2_cy7"      "2_cy5"      "2_alexa594"
## [12] "2_cy3"      "3_cy7"      "3_cy5"      "3_alexa594" "3_cy3"</code></pre>
</div>
<div id="establishing-global-image-parameters" class="section level3" number="3.2.2">
<h3>
<span class="header-section-number">3.2.2</span> Establishing global image parameters<a class="anchor" aria-label="anchor" href="#establishing-global-image-parameters"><i class="fas fa-link"></i></a>
</h3>
<p>As a first step, it is recommended that we determine the dynamic range of intensity values: i.e. the brightness value for below which (i.e. minimum brightness / floor) and above which (i.e. maximum brightness / ceiling) we can safely ignore gradation. A sensible minimum brightness would be intensities for background pixels where no tissue is imaged. For maximum brightness, we want a cap where all super-threshold intensities can effectively be considered ‘ON’. Ideally, these limits should be relatively consistent across FOVs.</p>
<p>We have created a function <code>getAnchorParams</code> that loads multiple FOVs and attempts to get global image parameters. As default, <code>brightness_max</code> and <code>brightness_min</code> are returned. For details, see the documentation for <code>getAnchorParams</code>.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/getAnchorParams.html">getAnchorParams</a></span><span class="op">(</span> nSamples <span class="op">=</span> <span class="fl">6</span> <span class="op">)</span></span></code></pre></div>
<p>These values are appended to <code>params</code>:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">params</span><span class="op">$</span><span class="va">brightness_min</span></span></code></pre></div>
<pre><code>##  [1] 0.02665484 0.07892208 0.02916114 0.04901184 0.03023053 0.06347031 0.02397402 0.04414353 0.02355671 0.07126881 0.02574727 0.04234602 0.02069601
## [14] 0.06184499 0.02228692 0.04452783</code></pre>
<p>Anchor FOVs are those used to derive global image parameters. These can be user-specified, or automatically determined with <code>getAnchorParams</code>. If automatic, <code>nSamples</code> number of FOVs are randomly selected. These can be visualised (in red) after running <code>getAnchorParams</code>:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">anchors</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "MMStack_1-Pos_001_003" "MMStack_1-Pos_005_000" "MMStack_1-Pos_007_009" "MMStack_1-Pos_008_009" "MMStack_1-Pos_009_001" "MMStack_1-Pos_002_003"</code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">anchors_plot</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-showanchors-1.png" width="672"></div>
<p>Random sampling should work for most cases. For more regular FOVs, users may instead opt for grid-based selection of FOVs: evenly spaced FOVs internal to the outermost ring of FOVs are selected. [Since we are re-obtaining anchor metrics, set <code>freshAnchors</code> to true.]</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/getAnchorParams.html">getAnchorParams</a></span><span class="op">(</span> anchorMode <span class="op">=</span> <span class="st">'grid'</span>, returnTroubleShootPlots <span class="op">=</span> <span class="cn">T</span>, freshAnchors <span class="op">=</span> <span class="cn">T</span> <span class="op">)</span></span></code></pre></div>
<p>And so this returns a different subset of anchors.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">anchors</span><span class="op">)</span></span></code></pre></div>
<pre><code>##  [1] "MMStack_1-Pos_001_001" "MMStack_1-Pos_001_003" "MMStack_1-Pos_001_005" "MMStack_1-Pos_001_007" "MMStack_1-Pos_003_002" "MMStack_1-Pos_003_004"
##  [7] "MMStack_1-Pos_003_006" "MMStack_1-Pos_003_008" "MMStack_1-Pos_005_001" "MMStack_1-Pos_005_003" "MMStack_1-Pos_005_005" "MMStack_1-Pos_005_007"
## [13] "MMStack_1-Pos_007_002" "MMStack_1-Pos_007_004" "MMStack_1-Pos_007_006" "MMStack_1-Pos_007_008"</code></pre>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">anchors_plot</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-showanchors_random_plot-1.png" width="672"></div>
<p><strong>Users should ensure that sufficient anchors are chosen for representative brightness caps and floors to be obtained.</strong> Additionally, different brightness caps / floors may be required for different samples, depending on how consistent imaging parameters were across samples. Inconsistent imaging parameters may be expected if samples were e.g. imaged on different days.</p>
<p>Additionally, should users wish to obtain other global parameters, this can be achieved by altering the <code>imageFunctions</code> and <code>summaryFunctions</code> fields in <code>getAnchorParams</code>.</p>
<p>For instance, say we wish to also obtain the mean brightness per image (<code>brightness_mean</code>). We first write a function to perform this (<code>custom_getMeanBrightness</code>), which will then be looped over every anchor FOV. We include this function among the named list <code>users_imageFunctions</code>, which we then input to <code>imageFunctions</code>. This returns a vector of mean brightness values for every anchor image, which are then summarised by the corresponding function in <code>summaryFunctions</code> (as default, we recommend <code>conservativeMean</code> – explained later – or some summary function robust to outliers).</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_getMeanBrightness</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span> <span class="va">im</span>, <span class="va">...</span> <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">meanBrightness</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span> <span class="va">im</span> <span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span> <span class="va">meanBrightness</span> <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">users_imageFunctions</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="st">'brightness_min'</span> <span class="op">=</span> <span class="va">imsBrightnessMin_MIP</span>, <span class="co">#MIP for Z-stacks</span></span>
<span>  <span class="st">'brightness_max'</span> <span class="op">=</span> <span class="va">imsBrightnessMax_MIP</span>,</span>
<span>  <span class="st">'brightness_mean'</span> <span class="op">=</span> <span class="va">custom_getMeanBrightness</span></span>
<span><span class="op">)</span></span>
<span><span class="va">users_summaryFunctions</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="st">'brightness_min'</span> <span class="op">=</span> <span class="va">conservativeMean</span>,</span>
<span>  <span class="st">'brightness_max'</span> <span class="op">=</span> <span class="va">conservativeMean</span>,</span>
<span>  <span class="st">'brightness_mean'</span> <span class="op">=</span> <span class="va">conservativeMean</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/getAnchorParams.html">getAnchorParams</a></span><span class="op">(</span> </span>
<span>  imageFunctions <span class="op">=</span> <span class="va">users_imageFunctions</span>, </span>
<span>  summaryFunctions <span class="op">=</span> <span class="va">users_summaryFunctions</span>,</span>
<span>  anchorMode <span class="op">=</span> <span class="st">'grid'</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>On <code>conservativeMean</code>: this is a function that takes the mean within a specific range, delimited by a certain number of standard deviations (default = 0.9). This should return mean values less skewed by outliers. An alternative could be <code>median</code>, or some user-defined function.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html">rgamma</a></span><span class="op">(</span> <span class="fl">1E4</span>, <span class="fl">2</span>, <span class="fl">1</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span> <span class="va">test</span>, breaks<span class="op">=</span><span class="fl">100</span>, main<span class="op">=</span><span class="st">'Histogram of values with mean (red) and conservativeMean (blue)'</span>, xlab<span class="op">=</span><span class="st">'Gamma distribution'</span> <span class="op">)</span>; </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span>, col<span class="op">=</span><span class="st">'red'</span><span class="op">)</span>; </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/conservativeMean.html">conservativeMean</a></span><span class="op">(</span><span class="va">test</span>, nSDs <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span>, col<span class="op">=</span><span class="st">'blue'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-conservativemean-1.png" width="672"></div>
<p>To evaluate the suitability of chosen thresholds, users can refer to the <code>troubleshootPlots</code> environment returned when <code>returnTroubleShootPlots</code> is true.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metric</span> <span class="op">=</span> <span class="st">'brightness_min'</span></span>
<span><span class="va">bit_number</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">troubleshootPlots</span><span class="op">[[</span> <span class="va">metric</span> <span class="op">]</span><span class="op">]</span><span class="op">[[</span> <span class="va">bit_number</span> <span class="op">]</span><span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-anchor-tsplots-1.png" width="960"></div>
<p>For each metric, we also report the observed thresholds as a quantile of each anchor image’s pixel intensity. These can be found in the <code>ANCHOR_QUANTILEQC.csv</code> file. We expect brightness floors to have a low quantile across FOVs (unless there are some empty FOVs included), and brightness ceilings to have high quantiles.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">out_dir</span>, <span class="st">'ANCHOR_QUANTILEQC.csv'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>##           metric bit MMStack_1.Pos_001_001 MMStack_1.Pos_001_003 MMStack_1.Pos_001_005 MMStack_1.Pos_001_007 MMStack_1.Pos_003_002 MMStack_1.Pos_003_004
## 1 brightness_min   1             0.9187895             0.1231919             0.2534933             0.2073274             0.2458487             0.5588561
## 2 brightness_max   1             0.9999800             0.9998100             0.9997967             0.9998654             0.9998590             0.9999914
## 3 brightness_min   2             0.9205798             0.1248851             0.2626584             0.2286615             0.2431207             0.5643622
## 4 brightness_max   2             0.9995589             0.9855621             0.9769110             0.9885266             0.9930865             0.9991272
## 5 brightness_min   3             0.9230520             0.1136280             0.2583598             0.1959322             0.2110254             0.4886617
## 6 brightness_max   3             1.0000000             1.0000000             0.9999945             1.0000000             0.9999952             1.0000000
##   MMStack_1.Pos_003_006 MMStack_1.Pos_003_008 MMStack_1.Pos_005_001 MMStack_1.Pos_005_003 MMStack_1.Pos_005_005 MMStack_1.Pos_005_007
## 1             0.5077519            0.10864865            0.08566187             0.4722320             0.6327399             0.1366108
## 2             0.9999982            0.99961480            0.99916499             0.9999821             1.0000000             0.9998951
## 3             0.5056256            0.13454552            0.08949598             0.4169337             0.4555937             0.1668553
## 4             0.9994388            0.99336617            0.87119601             0.9991592             0.9993902             0.9972692
## 5             0.4099523            0.09494156            0.08018513             0.3478341             0.3117073             0.1181132
## 6             1.0000000            0.99999726            0.99997740             1.0000000             1.0000000             0.9999995
##   MMStack_1.Pos_007_002 MMStack_1.Pos_007_004 MMStack_1.Pos_007_006 MMStack_1.Pos_007_008
## 1             0.1784223             0.4619367             0.1657847             0.9504158
## 2             0.9999053             0.9999768             0.9998435             0.9999966
## 3             0.1792000             0.5029208             0.1761314             0.9576270
## 4             0.9932534             0.9988799             0.9904303             0.9999356
## 5             0.1452180             0.4117146             0.1251103             0.9552260
## 6             1.0000000             1.0000000             0.9999977             1.0000000</code></pre>
</div>
</div>
<div id="maximum-intensity-projection" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> Maximum intensity projection<a class="anchor" aria-label="anchor" href="#maximum-intensity-projection"><i class="fas fa-link"></i></a>
</h2>
<p>For Z stacks, we default to maximum intensity projection (MIP) to collapse a 3D array into a 2D matrix. [If there are no other slices, then the 2D matrix is returned unchanged.]</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toy_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/array.html">array</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">11</span> <span class="op">/</span> <span class="fl">12</span>, nrow <span class="op">=</span> <span class="fl">3</span>, ncol <span class="op">=</span> <span class="fl">4</span>, byrow <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">11</span><span class="op">:</span><span class="fl">0</span> <span class="op">/</span> <span class="fl">12</span>, nrow <span class="op">=</span> <span class="fl">3</span>, ncol <span class="op">=</span> <span class="fl">4</span>, byrow <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="op">)</span>, dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/maxIntensityProject.html">maxIntensityProject</a></span><span class="op">(</span><span class="va">toy_example</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span> <span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span><span class="va">toy_example</span><span class="op">[</span>,,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">'Z slice 1'</span>, interpolate <span class="op">=</span> <span class="cn">F</span>, rescale <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span> <span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span><span class="va">toy_example</span><span class="op">[</span>,,<span class="fl">2</span><span class="op">]</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">'Z slice 2'</span>, interpolate <span class="op">=</span> <span class="cn">F</span>, rescale <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span> <span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span><span class="va">mip</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">'MIP'</span>, interpolate <span class="op">=</span> <span class="cn">F</span>, rescale <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-mip-1.png" width="672"></div>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Also note that similarly performs MIP (where possible) prior to finding minimum / maximum brightnesses.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">masmr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/masmr/man/imsBrightnessMin_MIP.html">imsBrightnessMin_MIP</a></span></span></code></pre></div>
<pre><code>## function (im, zDim = 3, ...) 
## {
##     if (length(dim(im)) != 2) {
##         im &lt;- suppressWarnings(maxIntensityProject(im, zDim = zDim))
##     }
##     result &lt;- imsBrightnessMin(im, ...)
##     return(result)
## }
## &lt;bytecode: 0x0000024257609d40&gt;
## &lt;environment: namespace:masmr&gt;</code></pre>
<p>Users wishing to process Z slices individually are encouraged refer to Troubleshooting (Chapter <a href="troubleshooting.html#troubleshooting">8</a>) for details.</p>
</div>
<div id="looping-through-images" class="section level2" number="3.4">
<h2>
<span class="header-section-number">3.4</span> Looping through images<a class="anchor" aria-label="anchor" href="#looping-through-images"><i class="fas fa-link"></i></a>
</h2>
<p>After establishing global parameters (e.g. brightness floors and ceilings), we are now ready to loop through our image data. It is strongly recommended that the user keeps track of the current FOV being processed in <code>params$current_fov</code> as follows:</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span> <span class="va">fovName</span> <span class="kw">in</span> <span class="va">params</span><span class="op">$</span><span class="va">fov_names</span> <span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="co">## Report current FOV / iteration </span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'\n'</span>, <span class="va">fovName</span>, <span class="st">': '</span>, </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">fov_names</span><span class="op">==</span><span class="va">fovName</span><span class="op">)</span>, <span class="st">' of '</span>, </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">fov_names</span><span class="op">)</span>, <span class="st">'...'</span> <span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## Check if file has been created</span></span>
<span>  <span class="kw">if</span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">resumeMode</span> <span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">checkFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">out_dir</span>, <span class="st">'SPOTCALL_'</span>, <span class="va">fovName</span>, <span class="st">'.csv.gz'</span><span class="op">)</span> <span class="op">)</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">checkFile</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">'FOV has been processed...Skipping...'</span><span class="op">)</span></span>
<span>      <span class="kw">next</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co">## Save the current_fov</span></span>
<span>  <span class="va">params</span><span class="op">$</span><span class="va">current_fov</span> <span class="op">&lt;-</span> <span class="va">fovName</span></span>
<span>  </span>
<span>  <span class="co">## Load image</span></span>
<span>  <span class="va">imList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/readImageList.html">readImageList</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">current_fov</span> <span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## If needed, MIP for each channel</span></span>
<span>  <span class="va">imList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/maxIntensityProject.html">maxIntensityProject</a></span><span class="op">(</span> <span class="va">imList</span> <span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## Image processing functions etc...</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div id="default-image-processing" class="section level3" number="3.4.1">
<h3>
<span class="header-section-number">3.4.1</span> Default image processing<a class="anchor" aria-label="anchor" href="#default-image-processing"><i class="fas fa-link"></i></a>
</h3>
<p>Designing custom image processing functions is a topic worthy of its own chapter (Chapter <a href="image-processing.html#image-processing">4</a>). For now, we will use the default processing provided in <code>masmr</code>:</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/getImageMetrics.html">getImageMetrics</a></span><span class="op">(</span> <span class="va">imList</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## </code></pre>
<p>The processed images are stored in a new <code>imMetrics</code> environment, and can be visualised as follows:</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span> <span class="va">imMetricName</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="va">imMetrics</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">imsub</span> <span class="op">&lt;-</span> <span class="va">imMetrics</span><span class="op">[[</span><span class="va">imMetricName</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span><span class="va">imsub</span><span class="op">[</span><span class="fl">1000</span><span class="op">:</span><span class="fl">1500</span>, <span class="fl">1000</span><span class="op">:</span><span class="fl">1500</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">)</span>, main <span class="op">=</span> <span class="va">imMetricName</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-imagemetrics-zoomed-1.png" width="672"></div>
</div>
<div id="image-registration" class="section level3" number="3.4.2">
<h3>
<span class="header-section-number">3.4.2</span> Image registration<a class="anchor" aria-label="anchor" href="#image-registration"><i class="fas fa-link"></i></a>
</h3>
<p>Images of the same FOV are rarely perfect overlaps of each other due to imprecise microscope movements. Additionally, optical distortions (e.g. chromatic aberration), may also cause images from different channels to become unaligned. It is thus highly recommended to align images before any downstream decoding (i.e. image registration).</p>
<p>It is up to the user to decide what image processing to perform to obtain images ideal for registration. We recommend a combination of low-pass and high-pass features: below, we combine images from <code>COARSE</code> and <code>DECODE</code> in <code>imMetrics</code> (i.e. band pass filtering).</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">forReg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">imList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">forReg</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/imNormalise.html">imNormalise</a></span><span class="op">(</span> <span class="va">imMetrics</span><span class="op">$</span><span class="va">COARSE</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">imMetrics</span><span class="op">$</span><span class="va">DECODE</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">forReg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">imList</span><span class="op">)</span> <span class="co">#Always keep the names from imList</span></span>
<span><span class="va">norm</span> <span class="op">&lt;-</span> <span class="va">forReg</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span> <span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span><span class="va">norm</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">'Band pass'</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## Warning in as.cimg.array(norm): Assuming third dimension corresponds to time/depth</code></pre>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span> <span class="fu">imager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imager/man/as.cimg.html">as.cimg</a></span><span class="op">(</span><span class="va">norm</span><span class="op">[</span><span class="fl">1000</span><span class="op">:</span><span class="fl">1500</span>, <span class="fl">1000</span><span class="op">:</span><span class="fl">1500</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">'Band pass (Zoom in)'</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-forregistration-1.png" width="672"></div>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Registration is accomplished with our <code>registerImages</code> function. The function works by maximising cross-correlation between two images (in Fourier transformed space). [Strictly, cross-correlation is first passed through a Gaussian filter centred around the origin, to upweight small shifts. Gaussian filter parameters can be tweaked using relevant parameters – see documentation for <code>registerImages</code>. Additionally, when an image stack is detected, maximum intensity projection is performed before registration.]</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/registerImages.html">registerImages</a></span><span class="op">(</span><span class="va">forReg</span>, returnTroubleShootPlots <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Warning in maxIntensityProject(imsub): 2D matrix provided...Skipping MIP...</code></pre>
<p>This function adds the necessary coordinate offsets to the <code>params</code> environment object. Shift vectors (and coordinates in general) are typically handled as complex numbers in <code>masmr</code>; with the real component reflecting the x-coordinate and the imaginary, y.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">params</span><span class="op">$</span><span class="va">shifts</span> <span class="op">)</span></span></code></pre></div>
<pre><code>##      0_cy7      0_cy5 0_alexa594      0_cy3      1_cy7      1_cy5 1_alexa594      1_cy3      2_cy7      2_cy5 2_alexa594      2_cy3      3_cy7 
##   0+0i   0+0i   0+0i   0-1i  -5+3i  -5+3i  -5+2i  -5+2i -12+5i -12+5i -11+4i -11+3i -14+2i
##      3_cy5 3_alexa594      3_cy3 
## -14+2i -14+3i -14+2i</code></pre>
<p>These offsets are added to existing coordinates so as to better align images. To evaluate the registration, users can set <code>returnTroubleShootPlots</code> in <code>registerImages</code> to true (which returns a <code>troubleshootPlots</code> object), or use <code>register_troubleshootPlots</code> directly (see documentation).</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">troubleshootPlots</span><span class="op">[[</span><span class="st">'REGISTRATION_EVAL'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'PRE_REGISTER'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span></span>
<span>  <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Before registration'</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-checkshifts-before-1.png" width="960"></div>
<p>The default is to plot registrations with respect to the reference FOV (red). A window centred on the brightest pixel in the reference FOV is selected as default.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">troubleshootPlots</span><span class="op">[[</span><span class="st">'REGISTRATION_EVAL'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'POST_REGISTER'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span></span>
<span>  <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'After registration'</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-checkshifts-after-1.png" width="960"></div>
<p>Additionally, the <code>registerImages</code> function also saves the reference bit image as a <code>REGIM_{ FOV Name }.png</code> file as part of its default behaviour. This can be called on for subsequent stitching.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">out_dir</span>, <span class="st">'REGIM_'</span>, <span class="va">params</span><span class="op">$</span><span class="va">current_fov</span>, <span class="st">'.png'</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="working-with-pixel-wise-metrics" class="section level2" number="3.5">
<h2>
<span class="header-section-number">3.5</span> Working with pixel-wise metrics<a class="anchor" aria-label="anchor" href="#working-with-pixel-wise-metrics"><i class="fas fa-link"></i></a>
</h2>
<p>Having aligned and processed our images, the next steps are to compile our images into a manageable data format, perform decoding, and successively filter out low quality pixels.</p>
<div id="moving-from-image-lists-to-a-dataframe" class="section level3" number="3.5.1">
<h3>
<span class="header-section-number">3.5.1</span> Moving from image lists to a dataframe<a class="anchor" aria-label="anchor" href="#moving-from-image-lists-to-a-dataframe"><i class="fas fa-link"></i></a>
</h3>
<p>The first step is to convert our images into a dataframe for easier use. Hereafter, decoding and filtering will be performed on this dataframe. This is accomplished with <code>consolidateImageMetrics</code>. Additionally, as the previous image lists are no longer needed, it is recommended users clean their environment with <code>cleanUp</code>.</p>
<p>If troubleshooting, users may wish to keep the <code>imMetrics</code> environment for the time being. However, do note that this tends to be a big object and may stretch memory requirements. [One solution, shown below, is to just keep a subset of the objects in <code>imMetrics</code>.]</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/consolidateImageMetrics.html">consolidateImageMetrics</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'Size of entire imMetrics: '</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="va">imMetrics</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="va">imMetrics</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1E-9</span>, <span class="st">' Gb'</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "Size of entire imMetrics: 5.04670648 Gb"</code></pre>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span> <span class="va">ix</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="va">imMetrics</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span> <span class="va">ix</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'DECODE'</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span> <span class="kw">next</span> <span class="op">}</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rm.html">remove</a></span><span class="op">(</span> list<span class="op">=</span><span class="va">ix</span>, envir<span class="op">=</span><span class="va">imMetrics</span> <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'Size of imMetrics after filtering: '</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="va">imMetrics</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="va">imMetrics</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1E-9</span>, <span class="st">' Gb'</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "Size of imMetrics after filtering: 1.121489776 Gb"</code></pre>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/cleanUp.html">cleanUp</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="st">'spotcalldf'</span>, <span class="co">#Necessary: must be specified, otherwise the dataframe will be deleted</span></span>
<span>  <span class="st">'imMetrics'</span>, <span class="co">#Optional: keep if want to make use of images in filterDF() later</span></span>
<span>  <span class="va">cleanSlate</span><span class="op">)</span> <span class="co">#Other things to keep</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The result is a dataframe with a unified coordinate system, in which pixels are rows and image metrics are columns.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span> <span class="va">spotcalldf</span> <span class="op">)</span>, format <span class="op">=</span> <span class="st">'html'</span>, booktabs <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-table"><table class="table table-sm">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
WX
</th>
<th style="text-align:right;">
WY
</th>
<th style="text-align:right;">
WZ
</th>
<th style="text-align:right;">
IDX
</th>
<th style="text-align:right;">
COARSE_01
</th>
<th style="text-align:right;">
DECODE_01
</th>
<th style="text-align:left;">
MASK_01
</th>
<th style="text-align:right;">
NORM_01
</th>
<th style="text-align:right;">
ORIG_01
</th>
<th style="text-align:right;">
COARSE_02
</th>
<th style="text-align:right;">
DECODE_02
</th>
<th style="text-align:left;">
MASK_02
</th>
<th style="text-align:right;">
NORM_02
</th>
<th style="text-align:right;">
ORIG_02
</th>
<th style="text-align:right;">
COARSE_03
</th>
<th style="text-align:right;">
DECODE_03
</th>
<th style="text-align:left;">
MASK_03
</th>
<th style="text-align:right;">
NORM_03
</th>
<th style="text-align:right;">
ORIG_03
</th>
<th style="text-align:right;">
COARSE_04
</th>
<th style="text-align:right;">
DECODE_04
</th>
<th style="text-align:left;">
MASK_04
</th>
<th style="text-align:right;">
NORM_04
</th>
<th style="text-align:right;">
ORIG_04
</th>
<th style="text-align:right;">
COARSE_05
</th>
<th style="text-align:right;">
DECODE_05
</th>
<th style="text-align:left;">
MASK_05
</th>
<th style="text-align:right;">
NORM_05
</th>
<th style="text-align:right;">
ORIG_05
</th>
<th style="text-align:right;">
COARSE_06
</th>
<th style="text-align:right;">
DECODE_06
</th>
<th style="text-align:left;">
MASK_06
</th>
<th style="text-align:right;">
NORM_06
</th>
<th style="text-align:right;">
ORIG_06
</th>
<th style="text-align:right;">
COARSE_07
</th>
<th style="text-align:right;">
DECODE_07
</th>
<th style="text-align:left;">
MASK_07
</th>
<th style="text-align:right;">
NORM_07
</th>
<th style="text-align:right;">
ORIG_07
</th>
<th style="text-align:right;">
COARSE_08
</th>
<th style="text-align:right;">
DECODE_08
</th>
<th style="text-align:left;">
MASK_08
</th>
<th style="text-align:right;">
NORM_08
</th>
<th style="text-align:right;">
ORIG_08
</th>
<th style="text-align:right;">
COARSE_09
</th>
<th style="text-align:right;">
DECODE_09
</th>
<th style="text-align:left;">
MASK_09
</th>
<th style="text-align:right;">
NORM_09
</th>
<th style="text-align:right;">
ORIG_09
</th>
<th style="text-align:right;">
COARSE_10
</th>
<th style="text-align:right;">
DECODE_10
</th>
<th style="text-align:left;">
MASK_10
</th>
<th style="text-align:right;">
NORM_10
</th>
<th style="text-align:right;">
ORIG_10
</th>
<th style="text-align:right;">
COARSE_11
</th>
<th style="text-align:right;">
DECODE_11
</th>
<th style="text-align:left;">
MASK_11
</th>
<th style="text-align:right;">
NORM_11
</th>
<th style="text-align:right;">
ORIG_11
</th>
<th style="text-align:right;">
COARSE_12
</th>
<th style="text-align:right;">
DECODE_12
</th>
<th style="text-align:left;">
MASK_12
</th>
<th style="text-align:right;">
NORM_12
</th>
<th style="text-align:right;">
ORIG_12
</th>
<th style="text-align:right;">
COARSE_13
</th>
<th style="text-align:right;">
DECODE_13
</th>
<th style="text-align:left;">
MASK_13
</th>
<th style="text-align:right;">
NORM_13
</th>
<th style="text-align:right;">
ORIG_13
</th>
<th style="text-align:right;">
COARSE_14
</th>
<th style="text-align:right;">
DECODE_14
</th>
<th style="text-align:left;">
MASK_14
</th>
<th style="text-align:right;">
NORM_14
</th>
<th style="text-align:right;">
ORIG_14
</th>
<th style="text-align:right;">
COARSE_15
</th>
<th style="text-align:right;">
DECODE_15
</th>
<th style="text-align:left;">
MASK_15
</th>
<th style="text-align:right;">
NORM_15
</th>
<th style="text-align:right;">
ORIG_15
</th>
<th style="text-align:right;">
COARSE_16
</th>
<th style="text-align:right;">
DECODE_16
</th>
<th style="text-align:left;">
MASK_16
</th>
<th style="text-align:right;">
NORM_16
</th>
<th style="text-align:right;">
ORIG_16
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
15055
</td>
<td style="text-align:right;">
255
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
14985
</td>
<td style="text-align:right;">
0.1685885
</td>
<td style="text-align:right;">
0.0666336
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0318499
</td>
<td style="text-align:right;">
0.0431527
</td>
<td style="text-align:right;">
0.1959525
</td>
<td style="text-align:right;">
0.0182777
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0874075
</td>
<td style="text-align:right;">
0.1050684
</td>
<td style="text-align:right;">
0.2910819
</td>
<td style="text-align:right;">
0.1395470
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.0616752
</td>
<td style="text-align:right;">
0.1104383
</td>
<td style="text-align:right;">
0.1467471
</td>
<td style="text-align:right;">
0.0036333
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0532875
</td>
<td style="text-align:right;">
0.0949858
</td>
<td style="text-align:right;">
0.0023963
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0242018
</td>
<td style="text-align:right;">
0.1435644
</td>
<td style="text-align:right;">
0.0209040
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0195158
</td>
<td style="text-align:right;">
0.0593125
</td>
<td style="text-align:right;">
0.1903854
</td>
<td style="text-align:right;">
0.0108405
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0055703
</td>
<td style="text-align:right;">
0.0601619
</td>
<td style="text-align:right;">
0.0732994
</td>
<td style="text-align:right;">
0.0026951
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0213105
</td>
<td style="text-align:right;">
0.1354508
</td>
<td style="text-align:right;">
0.0018844
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0410899
</td>
<td style="text-align:right;">
0.1167777
</td>
<td style="text-align:right;">
0.0052897
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0486866
</td>
<td style="text-align:right;">
0.2663939
</td>
<td style="text-align:right;">
0.0435117
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0160442
</td>
<td style="text-align:right;">
0.1010937
</td>
<td style="text-align:right;">
0.1288667
</td>
<td style="text-align:right;">
0.0019179
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0423075
</td>
<td style="text-align:right;">
0.1124755
</td>
<td style="text-align:right;">
0.0024016
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0433106
</td>
<td style="text-align:right;">
0.0792933
</td>
<td style="text-align:right;">
0.0028786
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0405763
</td>
<td style="text-align:right;">
0.1350443
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0568347
</td>
<td style="text-align:right;">
0.0570399
</td>
<td style="text-align:right;">
0.0021209
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0223294
</td>
</tr>
<tr>
<td style="text-align:left;">
15056
</td>
<td style="text-align:right;">
256
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
14986
</td>
<td style="text-align:right;">
0.1753839
</td>
<td style="text-align:right;">
0.0860337
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0431405
</td>
<td style="text-align:right;">
0.0468669
</td>
<td style="text-align:right;">
0.2066871
</td>
<td style="text-align:right;">
0.0489090
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0966803
</td>
<td style="text-align:right;">
0.1079719
</td>
<td style="text-align:right;">
0.2955125
</td>
<td style="text-align:right;">
0.1514781
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.0599694
</td>
<td style="text-align:right;">
0.1090731
</td>
<td style="text-align:right;">
0.1438671
</td>
<td style="text-align:right;">
0.0036322
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0511141
</td>
<td style="text-align:right;">
0.0923445
</td>
<td style="text-align:right;">
0.0023963
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0241780
</td>
<td style="text-align:right;">
0.1465902
</td>
<td style="text-align:right;">
0.0264190
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0257662
</td>
<td style="text-align:right;">
0.0613144
</td>
<td style="text-align:right;">
0.1964823
</td>
<td style="text-align:right;">
0.0234909
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0134243
</td>
<td style="text-align:right;">
0.0657939
</td>
<td style="text-align:right;">
0.0741845
</td>
<td style="text-align:right;">
0.0026951
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0200000
</td>
<td style="text-align:right;">
0.1455799
</td>
<td style="text-align:right;">
0.0018838
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0534705
</td>
<td style="text-align:right;">
0.1163260
</td>
<td style="text-align:right;">
0.0052897
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0529889
</td>
<td style="text-align:right;">
0.2757924
</td>
<td style="text-align:right;">
0.0654732
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0232150
</td>
<td style="text-align:right;">
0.1101432
</td>
<td style="text-align:right;">
0.1315909
</td>
<td style="text-align:right;">
0.0017231
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0464064
</td>
<td style="text-align:right;">
0.1207218
</td>
<td style="text-align:right;">
0.0024016
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0447938
</td>
<td style="text-align:right;">
0.0790790
</td>
<td style="text-align:right;">
0.0028778
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0397129
</td>
<td style="text-align:right;">
0.1441830
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0799452
</td>
<td style="text-align:right;">
0.0581769
</td>
<td style="text-align:right;">
0.0021209
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0198046
</td>
</tr>
<tr>
<td style="text-align:left;">
15057
</td>
<td style="text-align:right;">
257
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
14987
</td>
<td style="text-align:right;">
0.1767175
</td>
<td style="text-align:right;">
0.0890733
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0404089
</td>
<td style="text-align:right;">
0.0459683
</td>
<td style="text-align:right;">
0.2154453
</td>
<td style="text-align:right;">
0.0750645
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.1129332
</td>
<td style="text-align:right;">
0.1130610
</td>
<td style="text-align:right;">
0.2975720
</td>
<td style="text-align:right;">
0.1588251
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.0608004
</td>
<td style="text-align:right;">
0.1097382
</td>
<td style="text-align:right;">
0.1366940
</td>
<td style="text-align:right;">
0.0036309
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0454851
</td>
<td style="text-align:right;">
0.0879305
</td>
<td style="text-align:right;">
0.0023963
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0175016
</td>
<td style="text-align:right;">
0.1466302
</td>
<td style="text-align:right;">
0.0225324
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0132059
</td>
<td style="text-align:right;">
0.0572916
</td>
<td style="text-align:right;">
0.1976275
</td>
<td style="text-align:right;">
0.0270122
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0100954
</td>
<td style="text-align:right;">
0.0634068
</td>
<td style="text-align:right;">
0.0740865
</td>
<td style="text-align:right;">
0.0026951
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0204992
</td>
<td style="text-align:right;">
0.1562726
</td>
<td style="text-align:right;">
0.0018829
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0593196
</td>
<td style="text-align:right;">
0.1147730
</td>
<td style="text-align:right;">
0.0052897
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0534790
</td>
<td style="text-align:right;">
0.2846638
</td>
<td style="text-align:right;">
0.0895756
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0281234
</td>
<td style="text-align:right;">
0.1163376
</td>
<td style="text-align:right;">
0.1298415
</td>
<td style="text-align:right;">
0.0014963
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0440120
</td>
<td style="text-align:right;">
0.1246618
</td>
<td style="text-align:right;">
0.0024016
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0457579
</td>
<td style="text-align:right;">
0.0755643
</td>
<td style="text-align:right;">
0.0028770
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0333999
</td>
<td style="text-align:right;">
0.1513037
</td>
<td style="text-align:right;">
0.0000767
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0723273
</td>
<td style="text-align:right;">
0.0566595
</td>
<td style="text-align:right;">
0.0021209
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0192252
</td>
</tr>
<tr>
<td style="text-align:left;">
15058
</td>
<td style="text-align:right;">
258
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
14988
</td>
<td style="text-align:right;">
0.1725610
</td>
<td style="text-align:right;">
0.0745237
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0351278
</td>
<td style="text-align:right;">
0.0442310
</td>
<td style="text-align:right;">
0.2215627
</td>
<td style="text-align:right;">
0.0934877
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.1196076
</td>
<td style="text-align:right;">
0.1151508
</td>
<td style="text-align:right;">
0.2923042
</td>
<td style="text-align:right;">
0.1414761
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.0602755
</td>
<td style="text-align:right;">
0.1093181
</td>
<td style="text-align:right;">
0.1266107
</td>
<td style="text-align:right;">
0.0036295
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0360410
</td>
<td style="text-align:right;">
0.0870373
</td>
<td style="text-align:right;">
0.0023963
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0220558
</td>
<td style="text-align:right;">
0.1502949
</td>
<td style="text-align:right;">
0.0330563
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0283854
</td>
<td style="text-align:right;">
0.0621532
</td>
<td style="text-align:right;">
0.1963598
</td>
<td style="text-align:right;">
0.0262408
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0145165
</td>
<td style="text-align:right;">
0.0665772
</td>
<td style="text-align:right;">
0.0755687
</td>
<td style="text-align:right;">
0.0026951
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0216849
</td>
<td style="text-align:right;">
0.1567512
</td>
<td style="text-align:right;">
0.0018819
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0609281
</td>
<td style="text-align:right;">
0.1091272
</td>
<td style="text-align:right;">
0.0052897
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0472707
</td>
<td style="text-align:right;">
0.2918106
</td>
<td style="text-align:right;">
0.1114874
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0311911
</td>
<td style="text-align:right;">
0.1202091
</td>
<td style="text-align:right;">
0.1256519
</td>
<td style="text-align:right;">
0.0012436
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0408871
</td>
<td style="text-align:right;">
0.1283710
</td>
<td style="text-align:right;">
0.0024016
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0601454
</td>
<td style="text-align:right;">
0.0725193
</td>
<td style="text-align:right;">
0.0028761
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0309718
</td>
<td style="text-align:right;">
0.1566531
</td>
<td style="text-align:right;">
0.0003004
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0854233
</td>
<td style="text-align:right;">
0.0560639
</td>
<td style="text-align:right;">
0.0021209
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0190389
</td>
</tr>
<tr>
<td style="text-align:left;">
15059
</td>
<td style="text-align:right;">
259
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
14989
</td>
<td style="text-align:right;">
0.1633875
</td>
<td style="text-align:right;">
0.0446587
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0214699
</td>
<td style="text-align:right;">
0.0397380
</td>
<td style="text-align:right;">
0.2251548
</td>
<td style="text-align:right;">
0.1048259
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.1336188
</td>
<td style="text-align:right;">
0.1195380
</td>
<td style="text-align:right;">
0.2782481
</td>
<td style="text-align:right;">
0.0944740
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0467602
</td>
<td style="text-align:right;">
0.0985018
</td>
<td style="text-align:right;">
0.1213048
</td>
<td style="text-align:right;">
0.0036276
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0373857
</td>
<td style="text-align:right;">
0.0840205
</td>
<td style="text-align:right;">
0.0023963
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0242257
</td>
<td style="text-align:right;">
0.1538012
</td>
<td style="text-align:right;">
0.0442957
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0371955
</td>
<td style="text-align:right;">
0.0649749
</td>
<td style="text-align:right;">
0.1897434
</td>
<td style="text-align:right;">
0.0167363
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0093672
</td>
<td style="text-align:right;">
0.0628846
</td>
<td style="text-align:right;">
0.0774059
</td>
<td style="text-align:right;">
0.0026951
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0231669
</td>
<td style="text-align:right;">
0.1526377
</td>
<td style="text-align:right;">
0.0018807
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0517645
</td>
<td style="text-align:right;">
0.1032195
</td>
<td style="text-align:right;">
0.0052897
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0440939
</td>
<td style="text-align:right;">
0.2937106
</td>
<td style="text-align:right;">
0.1167685
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.0253241
</td>
<td style="text-align:right;">
0.1128049
</td>
<td style="text-align:right;">
0.1253053
</td>
<td style="text-align:right;">
0.0009766
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0451889
</td>
<td style="text-align:right;">
0.1312150
</td>
<td style="text-align:right;">
0.0024016
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0582171
</td>
<td style="text-align:right;">
0.0758693
</td>
<td style="text-align:right;">
0.0028751
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0411428
</td>
<td style="text-align:right;">
0.1587198
</td>
<td style="text-align:right;">
0.0004918
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0903877
</td>
<td style="text-align:right;">
0.0572023
</td>
<td style="text-align:right;">
0.0021209
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0208394
</td>
</tr>
<tr>
<td style="text-align:left;">
15060
</td>
<td style="text-align:right;">
260
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
14990
</td>
<td style="text-align:right;">
0.1521710
</td>
<td style="text-align:right;">
0.0107363
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
0.0325093
</td>
<td style="text-align:right;">
0.2237427
</td>
<td style="text-align:right;">
0.0974893
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.1340264
</td>
<td style="text-align:right;">
0.1196656
</td>
<td style="text-align:right;">
0.2608532
</td>
<td style="text-align:right;">
0.0406566
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0319765
</td>
<td style="text-align:right;">
0.0866704
</td>
<td style="text-align:right;">
0.1220588
</td>
<td style="text-align:right;">
0.0036248
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0383238
</td>
<td style="text-align:right;">
0.0772029
</td>
<td style="text-align:right;">
0.0023963
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0183838
</td>
<td style="text-align:right;">
0.1538350
</td>
<td style="text-align:right;">
0.0435356
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0308856
</td>
<td style="text-align:right;">
0.0629540
</td>
<td style="text-align:right;">
0.1776488
</td>
<td style="text-align:right;">
0.0024123
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
0.0498676
</td>
<td style="text-align:right;">
0.0800322
</td>
<td style="text-align:right;">
0.0026951
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0247582
</td>
<td style="text-align:right;">
0.1538761
</td>
<td style="text-align:right;">
0.0018794
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0594170
</td>
<td style="text-align:right;">
0.0982393
</td>
<td style="text-align:right;">
0.0052897
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0443117
</td>
<td style="text-align:right;">
0.2927845
</td>
<td style="text-align:right;">
0.1152722
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.0368280
</td>
<td style="text-align:right;">
0.1273229
</td>
<td style="text-align:right;">
0.1239065
</td>
<td style="text-align:right;">
0.0007128
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0407654
</td>
<td style="text-align:right;">
0.1332971
</td>
<td style="text-align:right;">
0.0024016
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0533966
</td>
<td style="text-align:right;">
0.0779187
</td>
<td style="text-align:right;">
0.0028741
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0433821
</td>
<td style="text-align:right;">
0.1544284
</td>
<td style="text-align:right;">
0.0006483
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0747240
</td>
<td style="text-align:right;">
0.0584646
</td>
<td style="text-align:right;">
0.0021209
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.0199081
</td>
</tr>
</tbody>
</table></div>
</div>
<div id="decoding" class="section level3" number="3.5.2">
<h3>
<span class="header-section-number">3.5.2</span> Decoding<a class="anchor" aria-label="anchor" href="#decoding"><i class="fas fa-link"></i></a>
</h3>
<p>Decoding involves calculating distances between image metric vectors and expected vectors from the codebook. As default, our <code>decode</code> function takes columns with <code>DECODE_XX</code> format and calculates</p>
<ul>
<li>
<code>COS</code>: cosine distances,</li>
<li>
<code>EUC</code>: Euclidean distances,</li>
<li>
<code>L2E</code>: L2-normalised Euclidean distances</li>
</ul>
<p>between image and codebook vectors. Both the nearest match and next nearest match (prefixed with <code>B</code>) are returned.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/decode.html">decode</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">)</span></span></code></pre></div>
<p>The relationship between the different distance metrics is illustrated below.</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span> plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="va">spotcalldf</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html">geom_hex</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">COS</span>, y<span class="op">=</span><span class="va">L2E</span><span class="op">)</span>, bins<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">'turbo'</span>, trans<span class="op">=</span><span class="st">'log10'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'top'</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="va">spotcalldf</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html">geom_hex</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">COS</span>, y<span class="op">=</span><span class="va">EUC</span><span class="op">)</span>, bins<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">'turbo'</span>, trans<span class="op">=</span><span class="st">'log10'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'top'</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="va">spotcalldf</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html">geom_hex</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">EUC</span>, y<span class="op">=</span><span class="va">L2E</span><span class="op">)</span>, bins<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">'turbo'</span>, trans<span class="op">=</span><span class="st">'log10'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'top'</span><span class="op">)</span></span>
<span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">1</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-distancerelations-1.png" width="672"></div>
<p>At this stage, pixels are assigned a preliminary gene label under the <code>g</code> column. As default, this is dependent on the nearest codebook match in cosine distance (column <code>COSLAB</code> in the dataframe).</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">'^g$|LAB$'</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, format <span class="op">=</span> <span class="st">'html'</span>, booktabs <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-table"><table class="table table-sm">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
g
</th>
<th style="text-align:right;">
COSLAB
</th>
<th style="text-align:right;">
EUCLAB
</th>
<th style="text-align:right;">
L2ELAB
</th>
<th style="text-align:right;">
BCOSLAB
</th>
<th style="text-align:right;">
BEUCLAB
</th>
<th style="text-align:right;">
BL2ELAB
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
15055
</td>
<td style="text-align:left;">
BLANK-NEW-05
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
15056
</td>
<td style="text-align:left;">
BLANK-NEW-05
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
</tr>
<tr>
<td style="text-align:left;">
15057
</td>
<td style="text-align:left;">
MKI67
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
15058
</td>
<td style="text-align:left;">
MKI67
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
15059
</td>
<td style="text-align:left;">
MKI67
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
51
</td>
<td style="text-align:right;">
51
</td>
<td style="text-align:right;">
51
</td>
</tr>
<tr>
<td style="text-align:left;">
15060
</td>
<td style="text-align:left;">
MKI67
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
51
</td>
<td style="text-align:right;">
51
</td>
<td style="text-align:right;">
51
</td>
</tr>
</tbody>
</table></div>
</div>
<div id="annotating-blanks" class="section level3" number="3.5.3">
<h3>
<span class="header-section-number">3.5.3</span> Annotating blanks<a class="anchor" aria-label="anchor" href="#annotating-blanks"><i class="fas fa-link"></i></a>
</h3>
<p>Users are recommended to filter out pixels with large distance metrics: i.e. those we cannot confidently decode. These thresholds can be determined <em>a priori</em>, or determined empirically by finding the threshold that separates blanks from non-blank calls.</p>
<p>There are several ways to define “blanks” that are provided in <code>masmr</code>:</p>
<ul>
<li>
<code>gBlank</code>: Whether the name of the gene assigned to a given pixel has the word “blank” in it.</li>
<li>
<code>bPossible</code>: Whether the next closest gene label for selected distance metrics are possible.</li>
<li>
<code>gbPossible</code>: Whether the next closest gene label for the default distance metric (<code>COS</code>) are possible.</li>
<li>
<code>consistentLabel</code>: Whether the gene labels agree across distance metrics.</li>
</ul>
<p>As default, <code>gBlank</code>, <code>gbPossible</code>, and <code>consistentLabel</code> tests are run. The number of tests failed is reported by <code>scoreBlanks</code>.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nBlankTestsFailed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/scoreBlanks.html">scoreBlanks</a></span><span class="op">(</span> <span class="va">spotcalldf</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">nBlankTestsFailed</span><span class="op">)</span></span></code></pre></div>
<pre><code>## nBlankTestsFailed
##       0       1       2       3 
## 2039428  616429  255396       2</code></pre>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">BLANK</span> <span class="op">&lt;-</span> <span class="va">nBlankTestsFailed</span> <span class="op">&gt;</span><span class="fl">0</span></span></code></pre></div>
<p>Say we decide that “blanks” are those that fail even one test. We can then examine the range of distance metrics between blanks and non-blanks. In general, distance metrics should be larger for blanks.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="va">spotcalldf</span>, measure.vars<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'COS'</span>, <span class="st">'EUC'</span>, <span class="st">'L2E'</span><span class="op">)</span><span class="op">)</span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">BLANK</span>, y<span class="op">=</span><span class="va">value</span>, colour<span class="op">=</span><span class="va">BLANK</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">BLANK</span>, y<span class="op">=</span><span class="va">value</span>, colour<span class="op">=</span><span class="va">BLANK</span><span class="op">)</span>, fill<span class="op">=</span><span class="st">'transparent'</span>, width<span class="op">=</span><span class="fl">0.25</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">variable</span>, scales<span class="op">=</span><span class="st">'free'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span> <span class="op">=</span> <span class="st">'red'</span>, <span class="st">'FALSE'</span> <span class="op">=</span> <span class="st">'black'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-blanksvnonblanks-distances-1.png" width="672"></div>
<p>A reasonable distance threshold might be the average of medians. This can be easily obtained with the <code>findThreshold</code> or <code>thresh_Quantile</code> function (see documentation for details).</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtout</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span> <span class="va">distanceMetric</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'COS'</span>, <span class="st">'EUC'</span>, <span class="st">'L2E'</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">thresh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/thresh_Quantile.html">thresh_Quantile</a></span><span class="op">(</span> <span class="va">spotcalldf</span><span class="op">[</span>,<span class="va">distanceMetric</span><span class="op">]</span>, label <span class="op">=</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">BLANK</span>, quantileFalse <span class="op">=</span> <span class="fl">0.5</span>, quantileTrue <span class="op">=</span> <span class="fl">0.5</span> <span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span> <span class="st">'\n'</span>, <span class="va">distanceMetric</span>, <span class="st">' threshold: '</span>, <span class="va">thresh</span> <span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">filtout</span> <span class="op">&lt;-</span> <span class="va">filtout</span> <span class="op">+</span> <span class="va">spotcalldf</span><span class="op">[</span>,<span class="va">distanceMetric</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">thresh</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code>## 
## COS threshold: 0.222254471692498
## EUC threshold: 3.30408867238758
## L2E threshold: 0.326703039218853</code></pre>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">filtout</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
##   FALSE    TRUE 
##  824288 2086967</code></pre>
</div>
<div id="filtering-and-evaluating-your-choice-of-filter" class="section level3" number="3.5.4">
<h3>
<span class="header-section-number">3.5.4</span> Filtering and evaluating your choice of filter<a class="anchor" aria-label="anchor" href="#filtering-and-evaluating-your-choice-of-filter"><i class="fas fa-link"></i></a>
</h3>
<p>How do we decide what is a good filter to apply to our data?</p>
<p>One idea is to use <em>a priori</em> knowledge of gene distributions in space. For instance, in our example dataset of an organoid, we expect dividing cells to be concentrated near the lumen. Accordingly, filtering should enrich lumenal <code>MKI67</code> (a marker for cell division).</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span> plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="va">spotcalldf</span><span class="op">[</span><span class="va">filtout</span><span class="op">&gt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">g</span><span class="op">==</span><span class="st">'MKI67'</span>,<span class="op">]</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html">geom_hex</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">WX</span>, y<span class="op">=</span><span class="va">WY</span><span class="op">)</span>, bins<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_c</a></span><span class="op">(</span>option<span class="op">=</span><span class="st">'turbo'</span>, trans<span class="op">=</span><span class="st">'log10'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span><span class="op">(</span> base_size <span class="op">=</span> <span class="fl">14</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Drop'</span><span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="va">spotcalldf</span><span class="op">[</span><span class="va">filtout</span><span class="op">==</span><span class="fl">0</span>  <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">g</span><span class="op">==</span><span class="st">'MKI67'</span>,<span class="op">]</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html">geom_hex</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">WX</span>, y<span class="op">=</span><span class="va">WY</span><span class="op">)</span>, bins<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_c</a></span><span class="op">(</span>option<span class="op">=</span><span class="st">'turbo'</span>, trans<span class="op">=</span><span class="st">'log10'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span><span class="op">(</span> base_size <span class="op">=</span> <span class="fl">14</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Keep'</span><span class="op">)</span></span>
<span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-spatialdistributionmki67-1.png" width="672"></div>
<p>Another option is to consider how filtering changes quality control metrics: e.g. a good filter should generally increase correlation between count profiles and previously obtained FPKM data, should decrease the percentage of blanks relative to true genes, etc. We provide a <code>filterDF</code> function that reports how a given filter changes these QC metrics (should relevant files be specified in <code>establishParams</code>).</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">params</span><span class="op">$</span><span class="va">verbose</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span> <span class="co">## To see output of filterDF, verbose is changed to True</span></span>
<span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/filterDF.html">filterDF</a></span><span class="op">(</span><span class="va">spotcalldf</span>, <span class="va">filtout</span><span class="op">&gt;</span><span class="fl">0</span>, returnTroubleShootPlots <span class="op">=</span> <span class="cn">T</span>, plotWindowRadius <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
## Applying filter: filtout &gt; 0...</code></pre>
<pre><code>## N pixels: 2911255 ( 100% ) --&gt; 824288 ( 28.31% )</code></pre>
<pre><code>## N blanks: 871827 ( 29.95% ) --&gt; 207062 ( 25.12% )</code></pre>
<pre><code>## LogFPKM Corr: 0.58239 --&gt; 0.59519</code></pre>
<pre><code>## LogN Probes Corr: 0.52337 --&gt; 0.51121</code></pre>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">params</span><span class="op">$</span><span class="va">verbose</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span> <span class="co">## Turn it back off if want to avoid messages entirely (warnings and errors will still be given)</span></span></code></pre></div>
<p>Perhaps the safest option – in that one that does not require <em>a priori</em> assumptions – is to perform some decoding by eye on a small FOV region, and ensure that one’s spotcalling pipeline returns results that agree with these manually obtained results. To attempt this, we provide a <code>returnTroubleShootPlots</code> parameter in <code>filterDF</code>. When set to <code>TRUE</code> (default <code>FALSE</code>), plots are returned showing pixel intensity images for each bit, overlaid with gene labels. [Do note that these plots are only available if users have kept their <code>imMetrics</code> environment, or have specified a list of images (<code>imList</code>).]</p>
<p>Plots show spots either <code>BEFORE</code> or <code>AFTER</code> filtering.</p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">troubleshootPlots</span><span class="op">[[</span><span class="st">'BEFORE'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-filter-QCBEFORE-1.png" width="864"></div>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">troubleshootPlots</span><span class="op">[[</span><span class="st">'AFTER'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-filter-QCAFTER-1.png" width="864"></div>
<p>Under the hood, <code>filterDF</code> makes use of the <code>spotcall_troubleshootPlots</code> function to return plots, and parameters for <code>spotcall_troubleshootPlots</code> can be input into <code>filterDF</code> to modify the plots returned. Alternatively, users may simply use <code>spotcall_troubleshootPlots</code> directly (see documentation).</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## An example</span></span>
<span><span class="va">plotList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/spotcall_troubleshootPlots.html">spotcall_troubleshootPlots</a></span><span class="op">(</span> </span>
<span>  <span class="va">spotcalldf</span>, </span>
<span>  plotWindowRadius <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  decodeMetric <span class="op">=</span> <span class="st">'DECODE'</span>, <span class="co">#Instead of DECODE, users may pick other objects in imMetrics</span></span>
<span>  chosenCoordinate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="fl">939</span> <span class="op">+</span> <span class="fl">340i</span>, <span class="co">#Specifying a custom coordinate</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">[</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'WX'</span>, <span class="st">'WY'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1i</span><span class="op">)</span><span class="op">)</span> <span class="co">#Specifying the first pixel in spotcalldf</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">plotList</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="va">plotList</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">p</span> <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><strong>Users are encouraged to optimise their image processing pipelines on one FOV, before applying the same pipelines to all other FOVs.</strong></p>
</div>
<div id="binarising-continuous-values" class="section level3" number="3.5.5">
<h3>
<span class="header-section-number">3.5.5</span> Binarising continuous values<a class="anchor" aria-label="anchor" href="#binarising-continuous-values"><i class="fas fa-link"></i></a>
</h3>
<p>Note that the current image vector contains continuous values. Users may wish to binarise values, such that superthreshold values are unambiguously assigned as ON, and subthreshold as OFF. To that end, we provide the <code>bitsFromDecode</code> function that searches for the optimal threshold that maximises accuracy of ON/OFF calls per bit. There are several ways to assay accuracy:</p>
<ul>
<li>
<code>precision</code>: True positives / (True positives + False positives)</li>
<li>
<code>recall</code>: True positives / (True positives + False negatives)</li>
<li>
<code>f1</code> (default): 2 * Precision * Recall / (Precision + Recall)</li>
<li>
<code>fb</code>: (1 + <span class="math inline">\(\beta^2\)</span>) * Precision * Recall / (<span class="math inline">\(\beta^2\)</span> * Precision + Recall)</li>
</ul>
<p>After thresholding, we can also calculate Hamming Distance from codebook vectors, which is stored under the <code>HD</code> column.</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/bitsFromDecode.html">bitsFromDecode</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">HD</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
##      0      1      2      3      4      5      6      7      8      9 
##  46125 288300 296934 124675  46694  15838   4772    851     94      5</code></pre>
<p>It is recommended to filter out pixels whose Hamming Distance to their assigned gene is &gt;1.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/filterDF.html">filterDF</a></span><span class="op">(</span><span class="va">spotcalldf</span>, <span class="va">spotcalldf</span><span class="op">$</span><span class="va">HD</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Binarisation is recommended if there is a clear cutoff in intensities between 0 vs 1 bits in the <code>DECODE</code> images. This assumption should be true in most cases, but may be violated if there are very different numbers of encoding probes for different genes: e.g. if there is a gene A that has 40 probes and a gene B with only 5 probes, ON for gene B will be dimmer than ON for gene A (and OFF for gene A may be brighter than OFF for gene B, depending on how thorough bleaching / washing was between imaging cycles). So a single intensity threshold separating ON vs OFF may not be appropriate here. <strong>In this scenario, users may wish to skip binarisation entirely.</strong></p>
</div>
<div id="obtaining-other-qc-metrics" class="section level3" number="3.5.6">
<h3>
<span class="header-section-number">3.5.6</span> Obtaining other QC metrics<a class="anchor" aria-label="anchor" href="#obtaining-other-qc-metrics"><i class="fas fa-link"></i></a>
</h3>
<p>Beyond image-codebook vector distances, we also provide other functions that users may find useful for performing further quality control checks. We provide several options for thresholding:</p>
<ul>
<li>
<code>thresh_Otsu</code>: Otsu thresholding. Useful for separating a bimodal distribution.</li>
<li>
<code>thresh_Elbow</code>: Identifies the elbow in a distribution. Useful for heavy tailed distributions.<br>
</li>
<li>
<code>thresh_Quantile</code>: Finds the average between quantiles in group A vs B. As default, the chosen quantiles are the medians of A and B.</li>
</ul>
<p>All thresholding functions are compatible with the <code>findThreshold</code> function.</p>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">F</span>, <span class="cn">T</span><span class="op">)</span>, <span class="fl">1E4</span>, <span class="cn">T</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">values</span><span class="op">[</span><span class="va">groups</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">groups</span><span class="op">)</span>, mean <span class="op">=</span> <span class="fl">1</span>, sd <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="va">values</span><span class="op">[</span><span class="op">!</span><span class="va">groups</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html">rgamma</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="op">!</span><span class="va">groups</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">1</span>, scale <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">values</span>, breaks<span class="op">=</span><span class="fl">1000</span>, main <span class="op">=</span> <span class="st">'Thresholds: Otsu (red), Elbow (blue), Quantile (green)'</span> <span class="op">)</span>; </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/findThreshold.html">findThreshold</a></span><span class="op">(</span><span class="va">values</span>, <span class="va">thresh_Otsu</span><span class="op">)</span>, col<span class="op">=</span><span class="st">'red'</span><span class="op">)</span>;</span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/findThreshold.html">findThreshold</a></span><span class="op">(</span><span class="va">values</span><span class="op">[</span><span class="va">values</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span><span class="op">]</span>, <span class="va">thresh_Elbow</span>, rightElbow <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, col<span class="op">=</span><span class="st">'blue'</span><span class="op">)</span>;</span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/findThreshold.html">findThreshold</a></span><span class="op">(</span><span class="va">values</span>, <span class="va">thresh_Quantile</span>, labels <span class="op">=</span> <span class="va">groups</span><span class="op">)</span>, col<span class="op">=</span><span class="st">'green'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-thresholding-syntheticexample-1.png" width="672"></div>
<div id="scoredeltaf" class="section level4 unnumbered">
<h4>
<code>scoreDeltaF</code><a class="anchor" aria-label="anchor" href="#scoredeltaf"><i class="fas fa-link"></i></a>
</h4>
<p>This function calculates <span class="math inline">\(\Delta F / F_0\)</span> metrics across image metrics. As default, all image metrics generated by <code>getImageMetrics</code> (and whose names are saved in <code>params$imageMetrics</code>) will be processed. [NB When <span class="math inline">\(F_0\)</span> is 0, <span class="math inline">\(\Delta F / F_0\)</span> is default set to the highest valid value for that column.]</p>
<p>Below, we show the <span class="math inline">\(\Delta F / F_0\)</span> alone, but users should read the <code>scoreDeltaF</code> documentation for the full list of metrics returned (e.g. <span class="math inline">\(F_1\)</span>, <span class="math inline">\(F_0\)</span> etc).</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/scoreDeltaF.html">scoreDeltaF</a></span><span class="op">(</span> <span class="va">spotcalldf</span> <span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span> <span class="va">spotcalldf</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">'^DFF0_'</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">)</span>, format <span class="op">=</span> <span class="st">'html'</span>, booktabs <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-table"><table class="table table-sm">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
DFF0_ORIG
</th>
<th style="text-align:right;">
DFF0_NORM
</th>
<th style="text-align:right;">
DFF0_COARSE
</th>
<th style="text-align:right;">
DFF0_DECODE
</th>
<th style="text-align:right;">
DFF0_MASK
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
15058
</td>
<td style="text-align:right;">
1.2802097
</td>
<td style="text-align:right;">
1.347051e+01
</td>
<td style="text-align:right;">
0.9599303
</td>
<td style="text-align:right;">
8.064703
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
15248
</td>
<td style="text-align:right;">
2.1340655
</td>
<td style="text-align:right;">
6.973976e+04
</td>
<td style="text-align:right;">
1.6488084
</td>
<td style="text-align:right;">
44.437034
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
15481
</td>
<td style="text-align:right;">
0.6950561
</td>
<td style="text-align:right;">
1.981764e+00
</td>
<td style="text-align:right;">
0.3761400
</td>
<td style="text-align:right;">
7.115631
</td>
<td style="text-align:right;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
15489
</td>
<td style="text-align:right;">
0.2565917
</td>
<td style="text-align:right;">
2.393276e+00
</td>
<td style="text-align:right;">
0.4356834
</td>
<td style="text-align:right;">
35.583974
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
15557
</td>
<td style="text-align:right;">
0.1342583
</td>
<td style="text-align:right;">
1.395903e-01
</td>
<td style="text-align:right;">
0.2187945
</td>
<td style="text-align:right;">
12.710667
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
15562
</td>
<td style="text-align:right;">
0.6546181
</td>
<td style="text-align:right;">
3.173345e+00
</td>
<td style="text-align:right;">
0.3342836
</td>
<td style="text-align:right;">
15.044078
</td>
<td style="text-align:right;">
11
</td>
</tr>
</tbody>
</table></div>
<p>We expect low confidence spots to on average have low <span class="math inline">\(\Delta F / F_0\)</span>, i.e. their ON bits are less distinguishable from their OFF bits.</p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">[</span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">F0_DECODE</span><span class="op">!=</span><span class="fl">0</span>,<span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">BLANK</span>, y<span class="op">=</span><span class="va">DFF0_DECODE</span>, colour<span class="op">=</span><span class="va">BLANK</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">BLANK</span>, y<span class="op">=</span><span class="va">DFF0_DECODE</span>, colour<span class="op">=</span><span class="va">BLANK</span><span class="op">)</span>, fill<span class="op">=</span><span class="st">'transparent'</span>, width<span class="op">=</span><span class="fl">0.25</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span> <span class="op">=</span> <span class="st">'red'</span>, <span class="st">'FALSE'</span> <span class="op">=</span> <span class="st">'black'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-deltaf-plot-1.png" width="672"></div>
</div>
<div id="spatialisadjacent" class="section level4 unnumbered">
<h4>
<code>spatialIsAdjacent</code><a class="anchor" aria-label="anchor" href="#spatialisadjacent"><i class="fas fa-link"></i></a>
</h4>
<p>The <code>spatialIsAdjacent</code> function searches in the immediate surroundings of a set of query coordinates. For instance, below, we ask if a given pixel is next to a blank. [Unsurprisingly, blanks will be reported as being next to blanks.]</p>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">isNextToBlank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/spatialIsAdjacent.html">spatialIsAdjacent</a></span><span class="op">(</span> <span class="va">spotcalldf</span>, <span class="st">'BLANK'</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/by.html">by</a></span><span class="op">(</span><span class="va">isNextToBlank</span>, <span class="va">spotcalldf</span><span class="op">$</span><span class="va">BLANK</span>, <span class="va">table</span><span class="op">)</span></span></code></pre></div>
<pre><code>## spotcalldf$BLANK: FALSE
## 
##  FALSE   TRUE 
## 240735  16688 
## ------------------------------------------------------------------------------------------------------------------ 
## spotcalldf$BLANK: TRUE
## 
##  TRUE 
## 77002</code></pre>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">NEXTTOBLANK</span> <span class="op">&lt;-</span> <span class="va">isNextToBlank</span></span></code></pre></div>
<p>In general, we are less confident about non-blank spots that <em>are</em> next to blanks, as evidenced by their higher codebook distances below.</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">BLANK</span>, <span class="op">]</span>, measure.vars<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'COS'</span>, <span class="st">'EUC'</span>, <span class="st">'L2E'</span><span class="op">)</span><span class="op">)</span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">NEXTTOBLANK</span>, y<span class="op">=</span><span class="va">value</span>, colour<span class="op">=</span><span class="va">NEXTTOBLANK</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">NEXTTOBLANK</span>, y<span class="op">=</span><span class="va">value</span>, colour<span class="op">=</span><span class="va">NEXTTOBLANK</span><span class="op">)</span>, fill<span class="op">=</span><span class="st">'transparent'</span>, width<span class="op">=</span><span class="fl">0.25</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">variable</span>, scales<span class="op">=</span><span class="st">'free'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span> <span class="op">=</span> <span class="st">'red'</span>, <span class="st">'FALSE'</span> <span class="op">=</span> <span class="st">'black'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Non-blank spots'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">'Is next to a blank'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-spatialisadj-plot-1.png" width="672"></div>
<p>Users may thus wish to “trim” pixel clusters, by removing non-blanks that are next to blanks: this is beneficial for clustering (described later). [Obviously, blanks should not be subject to the same filtering.]</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/filterDF.html">filterDF</a></span><span class="op">(</span><span class="va">spotcalldf</span>, <span class="va">spotcalldf</span><span class="op">$</span><span class="va">NEXTTOBLANK</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">BLANK</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="spatialhnndist" class="section level4 unnumbered">
<h4>
<code>spatialHNNDist</code><a class="anchor" aria-label="anchor" href="#spatialhnndist"><i class="fas fa-link"></i></a>
</h4>
<p>Another useful spatial metric is how far apart any two pixels with the same identity are from each other, here called “homotypic nearest neighbour (HNN) distance”. This can be calculated using <code>spatialHNNDist</code>.</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HNNDist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/spatialHNNDist.html">spatialHNNDist</a></span><span class="op">(</span> <span class="va">spotcalldf</span> <span class="op">)</span></span>
<span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">HNNDIST</span> <span class="op">&lt;-</span> <span class="va">HNNDist</span></span></code></pre></div>
<p>In general, blanks are less likely to be next to each other (possibly because of erroneous single pixel calls in noisy regions), while spots from the same gene tend to cluster - i.e. tend to have low HNN distances (below). Users typically do not have to perform thresholding on HNN distances given the later clustering step; though the distribution of distances can be helpful for determining minimum intracluster distances later on.</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hnnDistBins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cut.html">cut</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">HNNDIST</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="cn">Inf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">spotcalldf</span>, <span class="va">hnnDistBins</span><span class="op">)</span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_d</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">'HNN distance range (pixels)'</span>, option<span class="op">=</span><span class="st">'turbo'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">BLANK</span>, fill<span class="op">=</span><span class="va">hnnDistBins</span><span class="op">)</span>, position<span class="op">=</span><span class="st">'fill'</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'right'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">'Proportion'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-hnndistbins-1.png" width="672"></div>
</div>
<div id="scoreclassifier" class="section level4 unnumbered">
<h4>
<code>scoreClassifier</code><a class="anchor" aria-label="anchor" href="#scoreclassifier"><i class="fas fa-link"></i></a>
</h4>
<p>Finally, we have the <code>scoreClassifier</code> function. Under the hood, this function builds a classifier (logistic regression as default) to distinguish one class of pixels from another, using pixel-wise metrics in our dataframe. Users can specify covariates to include in the model; otherwise, all columns in the dataframe will be used.</p>
<p>Below is an example of predicting blanks from non-blanks. We exclude bit-specific columns (e.g. <code>DECODE_01</code>), covariates that are likely to highly correlate with others (e.g. <code>F1_DECODE</code> and <code>DFF0_DECODE</code> are likely to highly correlate), coordinate information, and any covariate that is equivalent to <code>BLANK</code>.</p>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">forExclusion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Xm'</span>, <span class="st">'Ym'</span>, <span class="st">'fov'</span>, <span class="st">'g'</span>, <span class="st">'WX'</span>, <span class="st">'WY'</span>, <span class="st">'IDX'</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">'_\\d+$|^DFF0_|^F1_|CV_|LAB$'</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">probabilityBLANK</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/scoreClassifier.html">scoreClassifier</a></span><span class="op">(</span><span class="va">spotcalldf</span>, <span class="st">'BLANK'</span>, variablesExcluded <span class="op">=</span> <span class="va">forExclusion</span><span class="op">)</span></span>
<span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">PBLANK</span> <span class="op">&lt;-</span> <span class="va">probabilityBLANK</span></span></code></pre></div>
<p>Accordingly the probability of being a blank is higher for blanks than non-blanks.</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">thresh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/thresh_Quantile.html">thresh_Quantile</a></span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">PBLANK</span>, <span class="va">spotcalldf</span><span class="op">$</span><span class="va">BLANK</span>, quantileFalse <span class="op">=</span> <span class="fl">0.5</span>, quantileTrue <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="va">spotcalldf</span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">BLANK</span>, y<span class="op">=</span><span class="va">PBLANK</span>, colour<span class="op">=</span><span class="va">BLANK</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">BLANK</span>, y<span class="op">=</span><span class="va">PBLANK</span>, colour<span class="op">=</span><span class="va">BLANK</span><span class="op">)</span>, fill<span class="op">=</span><span class="st">'transparent'</span>, width<span class="op">=</span><span class="fl">0.25</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">'Probability of being a blank'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span> <span class="op">=</span> <span class="st">'red'</span>, <span class="st">'FALSE'</span> <span class="op">=</span> <span class="st">'black'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept<span class="op">=</span><span class="va">thresh</span>, linetype <span class="op">=</span> <span class="st">'dashed'</span>, colour<span class="op">=</span><span class="st">'grey'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-blankprobs-visualise-1.png" width="672"></div>
<p>Thresholding and filtering can be performed using the classifier scores:</p>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/filterDF.html">filterDF</a></span><span class="op">(</span><span class="va">spotcalldf</span>, <span class="va">spotcalldf</span><span class="op">$</span><span class="va">PBLANK</span> <span class="op">&gt;</span> <span class="va">thresh</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div id="clustering" class="section level3" number="3.5.7">
<h3>
<span class="header-section-number">3.5.7</span> Clustering<a class="anchor" aria-label="anchor" href="#clustering"><i class="fas fa-link"></i></a>
</h3>
<p>Having performed QC to focus on pixels whose identity we are confident in, the next step is to cluster adjacent pixels with the same identity into a single punctum. This is accomplished with the <code>spatialClusterLeiden</code> function. Here, we want clusters at least 3 pixels big, and with a maximum distance between pixels of 2 pixels.</p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusterInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/spatialClusterLeiden.html">spatialClusterLeiden</a></span><span class="op">(</span> <span class="va">spotcalldf</span>, minNeighbours <span class="op">=</span> <span class="fl">3</span>, maxInterSpotDistancePixels <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">spotcalldf</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">clusterInfo</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">clusterInfo</span></span></code></pre></div>
<p>Users are recommended to visualise clusters across a small window. Note how specifying centroids only returns a single pixel coordinate per cluster.</p>
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Specify window to plot</span></span>
<span><span class="va">cxWindow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1200</span>, <span class="fl">1300</span>, <span class="fl">1200</span>, <span class="fl">1300</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Specify what genes assigned which colours</span></span>
<span><span class="co">## Manually:</span></span>
<span><span class="co"># genePalette &lt;- setNames(</span></span>
<span><span class="co">#   viridis::turbo(nrow(params$ordered_codebook)),</span></span>
<span><span class="co">#   rownames(params$ordered_codebook) )</span></span>
<span><span class="co">## Using previously established palette (established during prepareCodebook):</span></span>
<span><span class="va">genePalette</span> <span class="op">&lt;-</span> <span class="va">params</span><span class="op">$</span><span class="va">genePalette</span></span>
<span></span>
<span><span class="co">## Plot of labelled pixels vs centroids   </span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span> plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> </span>
<span>    <span class="va">spotcalldf</span><span class="op">[</span></span>
<span>      <span class="va">spotcalldf</span><span class="op">$</span><span class="va">WX</span> <span class="op">&gt;</span> <span class="va">cxWindow</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> </span>
<span>      <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">WX</span> <span class="op">&lt;</span> <span class="va">cxWindow</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> </span>
<span>      <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">WY</span> <span class="op">&gt;</span> <span class="va">cxWindow</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> </span>
<span>      <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">WY</span> <span class="op">&lt;</span> <span class="va">cxWindow</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="op">]</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span></span>
<span>      x<span class="op">=</span><span class="va">WX</span>, y<span class="op">=</span><span class="va">WY</span>, label<span class="op">=</span><span class="va">g</span>, </span>
<span>      colour<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">g</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">ordered_codebook</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span>      <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span> values<span class="op">=</span><span class="va">genePalette</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span>, </span>
<span>          plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span> limits<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">cxWindow</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">cxWindow</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_reverse</a></span><span class="op">(</span> limits<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">cxWindow</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="va">cxWindow</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'All pixels'</span><span class="op">)</span>,</span>
<span>  </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> </span>
<span>    <span class="va">spotcalldf</span><span class="op">[</span></span>
<span>      <span class="va">spotcalldf</span><span class="op">$</span><span class="va">CENTROID</span></span>
<span>      <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">WX</span> <span class="op">&gt;</span> <span class="va">cxWindow</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> </span>
<span>      <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">WX</span> <span class="op">&lt;</span> <span class="va">cxWindow</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> </span>
<span>      <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">WY</span> <span class="op">&gt;</span> <span class="va">cxWindow</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> </span>
<span>      <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">WY</span> <span class="op">&lt;</span> <span class="va">cxWindow</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="op">]</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span></span>
<span>      x<span class="op">=</span><span class="va">WX</span>, y<span class="op">=</span><span class="va">WY</span>, label<span class="op">=</span><span class="va">g</span>, </span>
<span>      colour<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">g</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">ordered_codebook</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span>      <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span> values<span class="op">=</span><span class="va">genePalette</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span>, </span>
<span>          plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span> limits<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">cxWindow</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">cxWindow</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_reverse</a></span><span class="op">(</span> limits<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">cxWindow</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="va">cxWindow</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Centroids only'</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">1</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-visualiseclustering-1.png" width="672"></div>
<p>Depending on what the user wants, the non-centroid pixels can be filtered. Additionally, users may wish to examine cluster sizes to find an appropriate range of pixels per cluster.</p>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maxClusterSize</span> <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span> <span class="va">spotcalldf</span><span class="op">[</span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">CENTROID</span>,<span class="op">]</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">CLUSTER_SIZE</span>, fill<span class="op">=</span><span class="va">BLANK</span><span class="op">)</span>, binwidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">BLANK</span>, ncol<span class="op">=</span><span class="fl">1</span>, scales<span class="op">=</span><span class="st">'free_y'</span>, strip.position <span class="op">=</span> <span class="st">'right'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">'Cluster size'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">'N centroids'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">maxClusterSize</span>, col<span class="op">=</span><span class="st">'grey'</span>, linetype<span class="op">=</span><span class="st">'dashed'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span> values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'FALSE'</span> <span class="op">=</span> <span class="st">'black'</span>, <span class="st">'TRUE'</span> <span class="op">=</span> <span class="st">'red'</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/spotcall-lookingatclustersizespercentroid-1.png" width="672"></div>
<p>As before, filtering is accomplished with <code>filterDF</code>:</p>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spotcalldf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/masmr/man/filterDF.html">filterDF</a></span><span class="op">(</span><span class="va">spotcalldf</span>, <span class="op">!</span><span class="op">(</span><span class="va">spotcalldf</span><span class="op">$</span><span class="va">CENTROID</span> <span class="op">&amp;</span> <span class="va">spotcalldf</span><span class="op">$</span><span class="va">CLUSTER_SIZE</span> <span class="op">&lt;</span> <span class="va">maxClusterSize</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div id="saving-spot-calls" class="section level2" number="3.6">
<h2>
<span class="header-section-number">3.6</span> Saving spot calls<a class="anchor" aria-label="anchor" href="#saving-spot-calls"><i class="fas fa-link"></i></a>
</h2>
<p>If we are satisified with our quality control, the final step is to use the <code>saveDF</code> function to save the dataframe. The default output is a <code>SPOTCALL_{ FOV Name }.csv.gz</code> file; users that have set <code>params$resumeMode</code> to <code>TRUE</code> may wish to first check if this file exists before proceeding with image processing.</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/masmr/man/saveDF.html">saveDF</a></span><span class="op">(</span> <span class="va">spotcalldf</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">out_dir</span>, <span class="st">'SPOTCALL_'</span>, <span class="va">params</span><span class="op">$</span><span class="va">current_fov</span>, <span class="st">'.csv.gz'</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Individual files are thereby generated for every FOV.</p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="introduction.html"><span class="header-section-number">2</span> Introduction</a></div>
<div class="next"><a href="image-processing.html"><span class="header-section-number">4</span> Image processing</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#spotcalling"><span class="header-section-number">3</span> Spotcalling</a></li>
<li>
<a class="nav-link" href="#pre-processing"><span class="header-section-number">3.1</span> Pre-processing</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#establishing-parameters"><span class="header-section-number">3.1.1</span> Establishing parameters</a></li>
<li><a class="nav-link" href="#getting-parameters-from-image-metadata"><span class="header-section-number">3.1.2</span> Getting parameters from image metadata</a></li>
<li><a class="nav-link" href="#preparing-the-codebook"><span class="header-section-number">3.1.3</span> Preparing the codebook</a></li>
<li><a class="nav-link" href="#house-keeping"><span class="header-section-number">3.1.4</span> House-keeping</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#preparing-to-loop-through-images"><span class="header-section-number">3.2</span> Preparing to loop through images</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#image-reading"><span class="header-section-number">3.2.1</span> Image reading</a></li>
<li><a class="nav-link" href="#establishing-global-image-parameters"><span class="header-section-number">3.2.2</span> Establishing global image parameters</a></li>
</ul>
</li>
<li><a class="nav-link" href="#maximum-intensity-projection"><span class="header-section-number">3.3</span> Maximum intensity projection</a></li>
<li>
<a class="nav-link" href="#looping-through-images"><span class="header-section-number">3.4</span> Looping through images</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#default-image-processing"><span class="header-section-number">3.4.1</span> Default image processing</a></li>
<li><a class="nav-link" href="#image-registration"><span class="header-section-number">3.4.2</span> Image registration</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#working-with-pixel-wise-metrics"><span class="header-section-number">3.5</span> Working with pixel-wise metrics</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#moving-from-image-lists-to-a-dataframe"><span class="header-section-number">3.5.1</span> Moving from image lists to a dataframe</a></li>
<li><a class="nav-link" href="#decoding"><span class="header-section-number">3.5.2</span> Decoding</a></li>
<li><a class="nav-link" href="#annotating-blanks"><span class="header-section-number">3.5.3</span> Annotating blanks</a></li>
<li><a class="nav-link" href="#filtering-and-evaluating-your-choice-of-filter"><span class="header-section-number">3.5.4</span> Filtering and evaluating your choice of filter</a></li>
<li><a class="nav-link" href="#binarising-continuous-values"><span class="header-section-number">3.5.5</span> Binarising continuous values</a></li>
<li><a class="nav-link" href="#obtaining-other-qc-metrics"><span class="header-section-number">3.5.6</span> Obtaining other QC metrics</a></li>
<li><a class="nav-link" href="#clustering"><span class="header-section-number">3.5.7</span> Clustering</a></li>
</ul>
</li>
<li><a class="nav-link" href="#saving-spot-calls"><span class="header-section-number">3.6</span> Saving spot calls</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>masmr: Modular Algorithms for Spotcalling in MERFISH in R</strong>" was written by Eugene Kwa. It was last built on 2025-07-18.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
